---
title: "Genome-wide DNA binding in HEPG2"
author: "JR"
date: "4/17/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Gviz)
library(ggpubr)
library(ggplot2)
library(tidyverse)
library(GenomicRanges)

source("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/util/intersect_functions.R")
source("util/plotting_functions.R")
source("util/_setup.R")
```

# Goal:

Here we aim to download all available DNA binding protein (DBP) profiles in a single cell state (measured by ChIP-seq) . This will allow us to investigate the binding properties of hundreds of DBPs in the same cellular context or background. We aim to address several questions: (i) What are the number of peaks and genome coverage for each DBP? (ii) What are the binding preferences for promoters, gene-bodies and intergenic genomic regions? (iii) What are the similarities and differences across DBPs based on their genome-wide binding profiles genome-wide? (iv) What properties or preferences do promoters have for binding events. (iv) Are there reservoir promoters in HepG2 as defined in k562 previously? (v) How does binding to a promoter affect the transcriptional output of that promoter?

To address these questions we have curated a set of X,000 ChIPs-eq data sets comprised of 486 DBPs in HEPG2 cells from the ENCODE consortrium. We required duplicate ChIP-seq experiments for a given DBP and other criterion that can be found here :

<https://www.encodeproject.org/report/?type=Experiment&status=released&assay_slims=DNA+binding&biosample_ontology.term_name=HepG2&assay_title=TF+ChIP-seq&biosample_ontology.classification=cell+line&files.read_length=100&files.read_length=76&files.read_length=75&files.read_length=36&assay_title=Control+ChIP-seq&assay_title=Histone+ChIP-seq&files.run_type=single-ended>

## These samples were selected on the following criteria:

1)  "chromatin" interaction data, then DNA binding data, cell line HEPG2, "TF-Chip-seq".
2)  We further selected "TF Chip-seq", "Control chip-seq" and "Histone Chip-seq".
3)  We selected several read lengths to get the most DNA binding proteins (DBPs)
4)  Read lengths: 100, 76, 75, 36
5)  ONLY SINGLE END READS (this eliminates 54 samples)

### Experimental data was downloading by (ENCODE report.tsv):

<https://www.encodeproject.org/report.tsv?type=Experiment&status=released&assay_slims=DNA+binding&biosample_ontology.term_name=HepG2&assay_title=TF+ChIP-seq&biosample_ontology.classification=cell+line&files.read_length=100&files.read_length=76&files.read_length=75&files.read_length=36&assay_title=Control+ChIP-seq&assay_title=Histone+ChIP-seq&files.run_type=single-ended>

### The FASTQ files were downloaded with:

"<https://www.encodeproject.org/metadata/?status=released&assay_slims=DNA+binding&biosample_ontology.term_name=HepG2&assay_title=TF+ChIP-seq&biosample_ontology.classification=cell+line&files.read_length=100&files.read_length=76&files.read_length=75&files.read_length=36&assay_title=Control+ChIP-seq&assay_title=Histone+ChIP-seq&files.run_type=single-ended&type=Experiment>"

MD5sums were checked with all passing (see encode_file_info function to reterive MD5Sum values that are not available from the encode portal (/util)

### Processing data:

We processed all the read alignments and peak calling using the NF_CORE ChIP-seq pipeline: (nfcore/chipseq v1.2.1)

## Next we created consensus peaks that overlap in both replicates

Our strategy was to take peaks in each replicate and find all overlapping peak windows. We then took the union length of the overlapping range in each peak window.

```{r creating-consensus-peaks}
# create_consensus_peaks requires an annotation .GTF file - loading in Gencode v32 annotations.
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/genomes/gencode.v32.annotation.gtf")

# Creating consensus peaks function to create a .bed file of overlapping peaks in each replicate.
# /util/intersect_functions.R

# TODO run this only on final knit
# create_consensus_peaks <- create_consensus_peaks(broadpeakfilepath = "/scratch/Shares/rinnclass/CLASS_2022/data/test_work/all_peak_files")

# exporting consensus peaks .bed files

# TODO run this only on final knit
# for(i in 1:length(consensus_peaks)) {
#  rtracklayer::export(consensus_peaks[[i]],
#                     paste0("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/consensus_peaks/",
#                             names(consensus_peaks)[i],
#                             "_consensus_peaks.bed"))
# }
```

# loading in consensus peaks to prevent rerunning create_consensus_peaks function

```{r consensus-peak-filtering}

# Loading in files via listing and rtracklayer import
consensus_fl <- list.files("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/consensus_peaks", full.names = T)

# importing (takes ~5min)
consensus_peaks <- lapply(consensus_fl, rtracklayer::import)

# cleaning up file names
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/consensus_peaks/|_consensus_peaks.bed","", consensus_fl)

# Filtering consensus peaks to those DBPs with at least 250 peaks
num_peaks_threshold <- 250
num_peaks <- sapply(consensus_peaks, length)
filtered_consensus_peaks <- consensus_peaks[num_peaks > num_peaks_threshold]

# Result: these were the DBPs that were filtered out.
filtered_dbps <- consensus_peaks[num_peaks < num_peaks_threshold]
names(filtered_dbps)

# We have this many remaining DBPs
length(filtered_consensus_peaks)

```

## Now we will determine the peak number and genome coverage for each DBP.

```{r peak number and coverage per DBP}

# Let's start with loading in the number of peaks each DBP has -- using length.
num_peaks_df <- data.frame("dbp" = names(filtered_consensus_peaks),
                           "num_peaks" = sapply(filtered_consensus_peaks, length))

# total genomic coverage of peaks for each dbp
num_peaks_df$total_peak_length <- sapply(filtered_consensus_peaks, function(x) sum(width(x)))

# Plotting distribution of peak number per dbp
hist(num_peaks_df$total_peak_length)

```

# Now we will create promoter annotations for lncRNA and mRNA and both.

We have created a funciton get_promter_regions that has up and downstream parameters

```{r promoter-gene-body-annotations}

# creating lncRNA and mRNA promoters
lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding", upstream = 1500, downstream = 1500))
names(lncrna_mrna_promoters) <- lncrna_mrna_promoters$gene_id
rtracklayer::export(lncrna_mrna_promoters, "analysis/results/lncRNA_mrna_promoters.gtf")

# creating lncRNAs promoter
lncrna_promoters <- get_promoter_regions(gencode_gr, biotype = "lncRNA", upstream = 1500, downstream = 1500) 
names(lncrna_promoters) <- lncrna_promoters$gene_id
rtracklayer::export(lncrna_promoters, "analysis/results/lncRNA_promoters.gtf")

# creating mRNA promoters
mrna_promoters <- get_promoter_regions(gencode_gr, biotype = "protein_coding", upstream = 1500, downstream = 1500)
names(mrna_promoters) <- mrna_promoters$gene_id
rtracklayer::export(lncrna_promoters, "analysis/results/mRNA_promoters.gtf")

# creating all genebody annotation
lncrna_mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                     gencode_gr$gene_type %in% c("lncRNA", "protein_coding")]
names(lncrna_mrna_genebody) <- lncrna_mrna_genebody$gene_id
rtracklayer::export(lncrna_mrna_genebody, "analysis/results/lncrna_mrna_genebody.gtf")

# creating lncRNA genebody annotation
lncrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                gencode_gr$gene_type %in% c("lncRNA")]
names(lncrna_genebody) <- lncrna_genebody$gene_id
rtracklayer::export(lncrna_mrna_genebody, "analysis/results/lncrna_genebody.gtf")

# creating mRNA genebody annotation
mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                              gencode_gr$gene_type %in% c("protein_coding")]
names(mrna_genebody) <-mrna_genebody$gene_id
rtracklayer::export(lncrna_mrna_genebody, "analysis/results/mrna_genebody.gtf")

```

# Determining the overlaps of chip peaks with promoters and genebodys

```{r DBP CHIP peaks overlap with promoters and genebody }

# creating index to subset lncRNA and mRNA annotations
lncrna_gene_ids <- lncrna_mrna_genebody$gene_id[lncrna_mrna_genebody$gene_type == "lncRNA"]
mrna_gene_ids <- lncrna_mrna_genebody$gene_id[lncrna_mrna_genebody$gene_type == "protein_coding"]

# using count peaks per feature returns number of annotation overlaps for a given DBP (takes ~5min)
promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, type = "counts")

# adding data to num_peaks_df
num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)
num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])

# gene body overlaps 
genebody_peak_counts <- count_peaks_per_feature(lncrna_mrna_genebody, 
                                                filtered_consensus_peaks, 
                                                type = "counts")


# adding data to num_peaks_df
num_peaks_df$peaks_overlapping_genebody <- rowSums(genebody_peak_counts)
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])
num_peaks_df$peaks_overlapping_mrna_genebody <- rowSums(genebody_peak_counts[,mrna_gene_ids])

write_csv(num_peaks_df, "analysis/results/num_peaks_df.csv")

```

# Plotting peak annotation features for DBPs

First we will plot the distribution of the number of peaks per DBP
```{r plotting peak annotation features}

num_peaks_df <- read_csv("analysis/results/num_peaks_df.csv")
# Distriubtion of peak numbers of all 460 DBPs
ggplot(num_peaks_df, aes(x = num_peaks)) + 
  geom_histogram(bins = 70)
summary(num_peaks_df$num_peaks)
```
The median number of peaks is nearly 16,000, but most DBPs tend to have many fewer binding events. The distribution looks like an expotential decay. Surprisingly, one DBP has nearly 150,000 peaks.

# Plotting the total genome coverage of each DBPs ChIP peaks

Here we want to see the relationship between the number of peaks and amount of genome covered
```{r num peaks versus genome coverage}

# Plotting number of peaks versus total genome coverage
ggplot(num_peaks_df, aes(x = num_peaks, y = total_peak_length)) +
  geom_point() + 
  geom_smooth(method = "gam", se = TRUE, color = "black", lty = 2)+
         
  ylab("BP covered") +
  xlab("Number of peaks") +
  ggtitle("Peak count vs. total bases covered")
ggsave("analysis/figures/peak_num_vs_coverage.pdf")
```
We observe a somewhat linear relationship between peak number and genome coverage
Thus, more peaks = more genome coverage. This could be interpreted as the peak sizes
are typically uniform and thus linearly increase genome coverage

# Plotting number of peaks versus overlaps with promoters

Here we are comparing the number of peaks for a given DBP 
to the total number of promoters those peaks overlap

```{r peak number versus promoter overlaps}

# Plotting number of peaks versus peaks overlapping promoters
ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_promoters)) +
  xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping promoters") +
  ggtitle("Relationship Between Number of DBP Peaks and Promoter Overlaps")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=FALSE, formula = 'y ~ x',
              color = "#a8404c") +
  stat_regline_equation(label.x = 35000, label.y = 18000) +
  ylim(0,60100) +
  xlim(0,60100)

ggsave("analysis/figures/3_peak_num_vs_promoter_coverage.pdf")
```
There is not a great linear fit across all the data. We observe that 
below 20,000 peaks there seems to be a linear relationship however
as the peak number increases fewer promoters are overlapping. This 
suggests that after 20,000 peaks most promoters are covered and now
the peaks are elsewhere.


# Peaks per DBP overlaps with genebody
Here we want to see how many of the peaks for each DBP
overlap genebodys. 

```{R peaks overlapping genebody}

# Plotting peak overlaps with genebody
ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_genebody)) +
  xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping genes") +
  ggtitle("Relationship Between Number of DBP Peaks and Gene Body Overlaps")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=F, formula = 'y ~ x',
              color = "#a8404c") +
  stat_regline_equation(label.x = 35000, label.y = 18000) +
  ylim(0,60100) +
  xlim(0,60100)

ggsave("analysis/figures/4_peak_num_vs_gene_body_coverage.pdf")
```
We observe that there is a very linear relationship between peak number and genebody overlap.
This means that the more peaks you have the more likely you are to overlap a genebody. One
way to think about this is that genebody overlaps explains most of the peaks -- or most peaks
overlap genebodys. We would expect dots (DBPs) that are below the line are binding outside genebodys.

However, one artifact could be if genebodys take up most of the genome this would also make a linear trend.

# Determing how much of the genome is comprised of genebodies.

```{r perentage of genome that is genebodys}

# there is a large amount of data explained (almost all by genebodys)
# Let's see what percentage of the genome genebodys cover:
reduced_gene_bodies <- gencode_gr[gencode_gr$type == "gene"] %>%
  GenomicRanges::reduce() %>%
  width() %>%
  sum()

# percentage of gene bodies in genome
reduced_gene_bodies/3.2e9

```
So we observe that genebodys do cover a lot of the genome. Since they 
represent a majority of the genome we would expect a linear relationship.
Nonetheless we conclude that most peaks overlap genebodys.


# Counting the number of overlaps at each promoter

promoters are the cols and DBPs rows thus we can retrieve the number of binding events at each promoter unlike the "counts parameter" that just gives total number of overlaps

```{r peak occurence matrix}

# Creating matrix of promoters(annotation feature) as cols and DBPs as rows (takes ~5min)
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, 
                                               type = "occurrence")

# test to make sure everything is in right order
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))

# Formatting final data.frame from peak occurence matrix
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "1.5_kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))
# exporting
write_csv(peak_occurence_df, "analysis/results/peak_occurence_dataframe.csv")

```


# Plotting
Here we are going to plot the distribution of the number of DBPs on a given promoter

```{r DBPs per promoter}

ggplot(peak_occurence_df, aes(x = number_of_dbp)) +
geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes") 

ggsave("figures/num_binding_events_per_promoter.pdf")

```
!COOL Result -- maybe !
We observe a bimodal distribution of binding at promoters.
This indicates that there are two types of promoters:
1) Those that have up to 70 or so DBPs bound
2) High binders that have more than 200 DBPs
We note a lag betweent 100-200 DBPs per promoter.
This could indicate that this is not a preferable number of DBPs.

# Multiple promoter window sizes

Let's iterate over promoter window sizes and see if this effect is maintained.
First we'll generate promoter windows of multiple 
```{r}
# Window sizes we want to iterate over.
window_size <- c(10, 30, 50, 100, 500, 1000, 1500, 2000, 2500, 3000)

# To test this function quickly, we'll subset gencode to a sampling of genes
gene_samples <- gencode_gr[gencode_gr$type == "gene"]
gene_samples <- gene_samples[gene_samples$gene_type %in% c("lncRNA", "protein_coding")]
gene_samples <- gene_samples[sample(1:length(gene_samples), 1000)]

promoter_list <- lapply(window_size, function(x) {
  GenomicRanges::promoters(gene_samples, upstream = x, downstream = x)
})
names(promoter_list) <- window_size

peaks_per_feature_list <- lapply(promoter_list, function(x) {
  count_peaks_per_feature(x, filtered_consensus_peaks, type = "occurrence")
})

num_dbp_list <- lapply(peaks_per_feature_list, function(x) {
  colSums(x)
})


for(i in 1:length(window_size)) {
  num_dbp_list[[i]] <- data.frame("gene_id" = names(num_dbp_list[[i]]),
                                  "num_dbp" = num_dbp_list[[i]],
                                  "window_size" = window_size[[i]])
}

num_dbp_df <- bind_rows(num_dbp_list)


ggplot(num_dbp_df, aes(x = num_dbp, color = window_size, group = window_size)) +
geom_density(alpha = 0.2) +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes") 
```
Awesome, we see that window size is not affecting the bimodal distribution.

# Binding properties versus total RNA expression.

```{r}

salmon_tpm <- read.csv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/18_running_RNAseq_NF_CORE/preclass_NF_core_RUN/results/salmon/salmon_merged_gene_tpm.csv")

samplesheet <- read_rds("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/19_rnaseq/final_samplesheet.rds")



g2s <- read_csv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/18_running_RNAseq_NF_CORE/g2s.csv") %>%
  dplyr::select(gene_id, gene_name)

tpm <- salmon_tpm %>% 
  pivot_longer(cols = 2:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samplesheet) %>%
  group_by(gene_id, condition) %>%
  summarize(tpm = mean(tpm, na.rm = T)) %>%
  filter(condition == "homo_sapiens_hepg2") %>%
  merge(g2s)


# Let's look at how many DBPs have a matching gene symbol in the gene expression dataset.
dbp_names <- tolower(num_peaks_df$dbp)
table(dbp_names %in% tolower(tpm$gene_name))
dbp_names[!(dbp_names %in% tolower(tpm$gene_name))]

# There are 14 missing. It looks like it's mostly the histone modifactions -- which are not genes.

dbp_expression <- tpm %>% filter(tolower(gene_name) %in% dbp_names) %>%
  mutate(gene_name = tolower(gene_name))



dbp_expression_vs_peak_number <- dbp_expression %>%
  left_join(num_peaks_df %>% dplyr::rename(gene_name = dbp) %>%
              mutate(gene_name = tolower(gene_name)))

ggplot(dbp_expression_vs_peak_number, aes(x = tpm, y = num_peaks)) +
  geom_point() +
  scale_y_log10() +
  xlim(0,50) +
  geom_smooth()
```

It seems that there's no correlation between the expression level of the gene and the number of peaks bound by the resulting protein. This makes sense due to the fact that mRNA levels and protein levels are only weakly correlated. 

```{r}
ggplot(dbp_expression_vs_peak_number, aes(x = tpm, y = total_peak_length)) +
  geom_point() +
    scale_y_log10() +
  xlim(0,50)
```

It seems that this is also true for the peak length (number of bases covered).


```{r}

promoter_features_df <- read.csv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/12_peak_features/peak_occurence_dataframe.csv")

promoter_features_df <- promoter_features_df %>%
  left_join(tpm)


ggplot(promoter_features_df, 
            aes(y = log2(tpm + 0.001), x = number_of_dbp, color = gene_type)) + 
geom_point(data = promoter_features_df %>% filter(tpm < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
 
  stat_cor() +
  # this is just to make it look nice.
  scale_x_continuous(expand = c(0,0)) +
  # adding colors manually
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  # title
  ggtitle("Expression vs. promoter binding events") + 
  # labeling axes
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 
```

Let's look at the distribution of expression values.

```{r}
ggplot(promoter_features_df, aes(x = tpm)) +
  geom_density() +
  xlim(0,2.5)
```
We see that most genes have very low expression and we see an inflection point in the density plot at around 0.25 TPM. Therefore, we'll call any gene with > 0.25 TPM as expressed.

```{r}
promoter_features_df <- promoter_features_df %>%
  mutate(expressed = tpm > 0.25)
table(promoter_features_df$expressed)
```

If we make this cutoff, 19,550 genes are expressed.

```{r}
ggplot(promoter_features_df %>% filter(!is.na(tpm)), aes(x = expressed)) +
  geom_bar() +
  geom_text(stat='count', aes(label=..count..), vjust=-1)
```

Let's look at the relationship between the number of TFs on a promoter and the expression only for expressed genes.

```{r}
ggplot(promoter_features_df %>% filter(expressed == TRUE), 
            aes(y = log2(tpm), x = number_of_dbp, color = gene_type)) + 
  geom_bin_2d() +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor()
```

For expressed genes (TPM > 1), there is still a correlation between higher numbers of DBPs on the promoter and the expression level. This trend carries through even very high numbers of DBPs (> 200). And more so for mRNAs than lncRNAs.

Suprisingly, there are genes which have no DBPs bound in this dataset, but are still expressed. What are the gene types of these genes? Are they transcribed by a different polymerase? Perhaps this highlights the fact that although we have a large number of DBPs in this dataset, there are still many other transcription factors not profiled. We have approximately 1/3 of the total identified human TFs (460/1200).

```{r}
ggplot(promoter_features_df %>% filter(number_of_dbp == 0, !is.na(tpm)), aes(x = expressed)) +
  geom_bar() +
  geom_text(stat='count', aes(label=..count..), vjust=-1) 
```
This is probably explained that there is a different TSS that is bound -- we are using the longest isoform
but this may not be the one that is active in this data. A better approach would be to do this on an isoform
level.  

## ATAC-seq 

Let's look to see if there are differences in the number of ATAC-seq peaks over the promoters.

```{r}
# wget https://www.encodeproject.org/files/ENCFF536RJV/@@download/ENCFF536RJV.bed.gz
atac_peaks <- read.table("data/ENCFF536RJV.bed.gz",
                         col.names = c("chrom", "start", "end", "name", "score", 
                                       "strand", "signal_value", "pval", 
                                       "qval", "summit"))

atac_peaks_gr <- GRanges(seqnames = atac_peaks$chrom,
                         ranges = IRanges(start = atac_peaks$start,
                                          end = atac_peaks$end))
atac_peaks_list <- list(atac_peaks_gr)
names(atac_peaks_list) <- "atac"
# Look for overlaps with promoters
atac_promoter_ov <- count_peaks_per_feature(lncrna_mrna_promoters, atac_peaks_list, type = "counts")
atac_promoter_df <- atac_promoter_ov %>% 
  t() %>%
  as.data.frame() %>%
  rownames_to_column("gene_id")

ggplot(atac_promoter_df, aes(x = atac)) +
  geom_density(adjust = 1.9)

promoter_features_df <- promoter_features_df %>%
  left_join(atac_promoter_df)

ggplot(promoter_features_df %>% filter(number_of_dbp > 100), aes(x = number_of_dbp, y = atac)) +
  geom_bin_2d() +
  geom_smooth() +
  stat_cor()

ggplot(promoter_features_df %>% filter(number_of_dbp > 100), aes(x = factor(atac), y = number_of_dbp)) +
  geom_violin()

```


