---
title: "UCSC custom tracks"
author: "JR"
date: "10/6/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("../../../util/intersect_functions.R")
```


The goal today is to learn more about our "peak files" and
data formatting (typical first steps)

The NF_CORE pipeline gave us a bunch of "peak files" that contain
information on where each peak for each DBP resides in the genome.

These files are here:

```{bash}

cd /scratch/Shares/rinnclass/CLASS_2022/data/peaks

# Let's see how many there are:

cd .. | ls | wc -l

```

Ok so let's go over the information in the files and then format them for UCSC.

ALL .BED require the first three fields to be :

```
Chr start stop
```

This is our first step towards understanding Iranges or Genomic Ranges. 
These files will always be listed in "intervals" or ranges of features.

Let's take a look at the first line of MACS broad.Peaks file:

```
chr1	827072	827564	ASH2L_R1_peak_12	61	.	5.22849	8.708	6.12277
```

So we see this peak is on chromosome 1 and the range of the peaks starts at 
827072 and peak end range is 827564. Those are minimal requirements for a .bed file.

We see additional information here now:

```
Column 4: peak name
Column 5: peak score  #int(-10*log10pvalue) 
Column 6: strand
Column 7: fold change at peak summit
column 8: -log10pvalue at peak summit (pval)
column 9 -log10qvalue at peak summit (qval)
```


Now we need to do some formatting to get the file into UCSC which loves standard Bed3
but get's confused if there is more information and finally a header.

There are lots of unfinished regions of the genome that have "contig" names (GL000218.1)
we want to remove these as UCSC doesn't recognize them (and they are typically ignored -hmmm)

They look like this:

```
GL000218.1	97231	97515	ASH2L_R1_peak_1	36	.	4.45278	6.12847	3.69243
GL000219.1	99163	100121	ASH2L_R1_peak_2	48	.	3.60875	7.41229	4.88736
GL000251.2	2243479	2243999	ASH2L_R1_peak_3	18	.	3.61082	4.06835	1.81115
GL383563.3	324323	324616	ASH2L_R1_peak_4	45	.	4.90069	7.02559	4.52199
KI270729.1	160787	161306	ASH2L_R1_peak_5	104	.	6.52922	13.3587	10.4542
KI270830.1	16789	17306	ASH2L_R1_peak_6	17	.	3.52976	3.99846	1.75013
```

Well this sounds simple but we need several steps to get there:

First make a list of file names with list.files. 
```{r}

fl <- list.files("/scratch/Shares/rinnclass/CLASS_2022/data/peaks", full.names = T)
summary(fl)

# table will let us see what these files are labeled as
table(fl)

# we see there is the file path appended onto the file name
# we can also see that the DBP name is the filename too.
# let's clean up the label  with 'gsub' function

```


# gsub: cleaning up file names 
```{R cleaning up file names}

?names
# by default the row names are 1-how many rows
# however we can name these with names
# note that "names" are a meta label and don't contribute in a column

?gsub
# gsub is just like find and replace 
# syntax: what do you want to replace "|" means 'or'

# gsub to remove file path and _peak.broadPeak

names(fl) <- gsub("/scratch/Shares/rinnclass/CLASS_2022/data/peaks/|_peaks.broadPeak", "", fl)

# Let's look at it 

summary(fl)
table(fl)
names(fl)
```

# making bed format for each peak file
Now we are going to make a table of the peak files -- 
then we can print them out with header next
```{R creating a specific .bed file format}

# to read in a bed file we can use read.table
?read.table

# lapply will run a function on a list of files
?lapply

# running lapply to use function 'read.table' to get the values in each peak file 
# note fl is just file path list

peaks <- lapply(fl[1:3], read.table, col.names = c("chr", "start", "end", "name", "score", "strand", "signal_value", "pval", "qval"))

# let's take a look
summary(peaks)
names(peaks)
fl[1:3] 

# note that read.table is still using the file path to read in each file 
# and names is a meta-header

```


# filter out the non-canonical chromosomes
```{R}

# Make a object of the chromosomes we want to keep.

canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")

# now let's use lapply to run a generic function 'function(x)'
# we run function(x) and then send x to filter (using pipe %>%)
# filter uses %in% to select only chromosomes in 'canonical_chr'
  
peaks <- lapply(peaks, function(x) x %>% filter(chr %in% canonical_chr))


```

# creating a header for UCSC

Next we need a header to be compatible with UCSC 
We are using a broadPeak file which UCSC recognizes
but we need to tell it which one it is in header

There is a lot of infomation here on file formats for UCSC:

https://genome.ucsc.edu/FAQ/FAQformat.html#format1

The header we need is : track type=broadPeak name="
```{R adding header to peak files}

# we are using paste0 to print the header text
# then we are adding the names as the value the header equals with 'names' function

headers <- paste0("track type=broadPeak name=", names(fl)[1:3])

# we are going to prepare the file names for a for loop to write them all out.
# however we are only indexing the first 3 files in fl (fl[1:3])

new_filenames <- paste0("/scratch/Shares/rinnclass/CLASS_2022/data/ucsc/broadPeak/", names(fl)[1:3], ".broadPeak")

new_filenames
```

Cool, now we have an object "new_filenames" that has 
the file path we want to write out each file to.
Let's use a for loop to print each file out separately and append the header.
```{R}
# let's do so more indexing and only print out the first two of the three files.

for(i in 1:length(peaks[1:2])) {
  # Write the header line
  writeLines(headers[[i]], new_filenames[[i]])
  # Append the broadPeak table data
  
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}

# note append=TRUE is very important otherwise when the peak data is imported it will overwrite header
```

Cool now we have these files printed out and can be uploaded to UCSC genome browser.

# Now move to local computer and upload to UCSC custom tracks

# if there is an issue I have hosted them on AWS here:

https://class2022.s3.us-east-2.amazonaws.com/ADNP_R2.broadPeak
https://class2022.s3.us-east-2.amazonaws.com/ADNP_R1.broadPeak

Let's go to the browser and load them in!

We also have a track hub of all peaks and BigWig files.

https://genome.ucsc.edu/s/BCHM_5631/firre_promoter

*********************
EXCERCISE
*********************

Create a UCSC peak file with addition features in the header.

Hint:

track type=broadPeak visibility=3 db=hg19 name="bPk" description="ENCODE broadPeak Example"
browser position chr1:798200-800700
chr1     798256 798454 .       116      .       4.89716 3.70716 -1
chr1     799435 799507 .       103      .       2.46426 1.54117 -1
chr1     800141 800596 .       107      .       3.22803 2.12614 -1

