---
title: "Genome Features as Ranges"
author: "JR"
date: "10/6/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(rtracklayer)
library(tidyverse)
library(ggpubr)
library(Gviz)
source("util/_setup.R")
source("util/intersect_functions.R")
source("util/plotting_functions.R")
```

### Have you ever wanted to find overlaps between two lists of genes? Well
Iranges and GRanges are just the thing!

#### Iranges allows one to index "windows" with properties. Say you had a vector
of all genes and wanted to track start and stop. Iranges can keep the start, 
stop and width of values. So this list of ranges (start and stop) for each 
gene-length is essentially a vector or each column a gene id. With this indexing
we can now apply several attributes to these ranges. For example if an enhancer 
or other feature resided in the gene we could also track the start, stop of that
"interval". 

#### Granges is even more specially aadapted for features in a given genome. All
the start, stops for each feature is given similar to Iranges. However, we could
have a big problem since we have 23 chromosomes in the geneome! So simple start
and stop indexing isn't enough as there could be up to 23 redudant "intervals" 
that are actually unique. So we need to consider choromosome and coordinates. 
Worse we also should think about the strand. So GRanges works similar to Iranges
just having the advantage of indexing chromosome and strand to the start and stop.


#### Let's go explore the human geneome annotations from GENCODE.
We will do this by importing a GTF file of these annotaitons.
A GTF file is similar in a way to a .bed file with start stop of features
and other associated metadata that is described here:

https://uswest.ensembl.org/info/website/upload/gff.html


## Ok enough background let's start exploring the features of the human genome!
```{r}

gencode_gr <- rtracklayer::import("data/gencode.v32.annotation.gtf")

# we used rtracklayer::import to turn the annotations into start and stops (Iranges) w/ chr (Granges)

```

Now let's look at the gene_id indexable part of this GRanges
```{r}

?seqnames
  
seqnames(gencode_gr) %>% head()

# we can see that "levels" are the chromosome names and are a meta-level
```

let's get more info with table to summary using tidy verse %>%
```{R}
table(seqnames(gencode_gr)) %>% 
  summary()
# this tells us how many features there are in the genome.

table(seqnames(gencode_gr)) %>%
  head()
# This is telling us how many features are associated with each chromosome.

# Let's find out how many features the mitochondrial chr has (25th level)
table(seqnames(gencode_gr))[25] 

# Let's look in the environment at what is in @metadata that we can index into with $

```

Now let's dig in a little deeper

```{R}
table(gencode_gr@elementMetadata$gene_id) %>% 
  head()
# This tell us how many features are associated with each gene id.

# Let's see what we get with gene_name
table(gencode_gr@elementMetadata$gene_name)

# it's basically memorization of what is a gene name or gene id 

```

#TODO
Let's analyze some of the features of our genome!

```{r}

#let's organize the data frame first and put it into a data frame.

gencode_gr_length <- data.frame("gene_id" = gencode_gr@elementMetadata$gene_id,
                           "gene_name" = gencode_gr@elementMetadata$gene_name,
                           "gene_type" = gencode_gr@elementMetadata$gene_type,
                           "type" = gencode_gr@elementMetadata$type,
                           "start" = gencode_gr@ranges@start,
                           "width" = gencode_gr@ranges@width)
                           
```


Ok so now we have a new data frame with the information we want from gencode_gr
Note we see that we can index the RANGES for each gene with @ranges "level"

Let's look at the width for mRNA genes
```{R}

mrna_df <- filter(gencode_gr_length, gene_type == "protein_coding", type == "gene")

# we can easily plot a histogram of mRNA gene widths.

hist(log10(mrna_df$width), breaks = 60)

# same for lncRNAs

lncrna_df <- filter(gencode_gr_length, gene_type == "lncRNA", type == "gene")

hist(log10(lncrna_df$width), breaks = 60)

```

Now let's go find the widths associated with gene-length.
```{R}

table(gencode_gr@elementMetadata$type)

## cool this tell us a lot about the genome with one line! How many genes are

```

let's find the longest gene in the genome!
```{R} 

# Let's see the range of gene sizes

summary(width(gencode_gr))

# Let's turn gencode_gr into a data.frame and filter to only genes.

gencode_df <- gencode_gr %>%
  as.data.frame() %>%
  filter(type == "gene")

# Note the difference in observations goes from millions to ~60K


# Let's find the smallest genes
gencode_df %>% 
  arrange(width) %>%
  head()

# Now let's find the Largest genes
gencode_df %>%
  arrange(-width) %>%
  head()


# We can also store the longest gene as an object

longest_gene <- gencode_df %>%
  filter(width == max(width))


# let's take a look at what information we have for this gene:

longest_gene

```

Now let's find the total amount of the genome covered by exons.
We start by defining total_exon_length of all exons in genome.

# we need to use reduce to collapse all the exons that overlap into the longest exon
# Exon 1.1  _____________
# Exon 1.2      _____________
# reduce    _________________
# intersect     _________


Ok let's find out what % of genome is exons!
```{R}

total_exon_length <- gencode_gr[gencode_gr$type == "exon"] %>%
  GenomicRanges::reduce() %>% 
  width() %>%
  sum()

summary(total_exon_length)
# We now have the total bases covered by exons in the human genome. We also know
# the human genome is 3.9B base pairs so let's find the percentage by division:

total_exon_length/3.2e9
# Boom we can see the human genome is about 4% exons!

```



Now let's use GRanges to find overlaps of genome features and peaks from ChIPseq.
This is probably one of the most commonly used functions in genome sciences!

```{r}
### Let's import POLR2A Chip-seq files from k562 ENCODE project.

pol2_chip <- rtracklayer::import("data/POLR2A_consensus_peaks_filter.bed")
```

#TODO
We'll need to define the promoters using the handy promoters function in GRanges
```{R}
?promoters

# let's add 3Kb upstream and downstream from the TSS to define "promoters"
gencode_promoters <- promoters(gencode_gr[gencode_gr$type == "gene"], 
                               upstream = 3e3, 
                               downstream = 3e3)


length(gencode_promoters)

```


#TODO Understand these numbers
Now we have promoter ranges and peak ranges let's use findOverlaps function



Reality is:

 1     2     3 
12831   464     4 
> 12831+464+464+4+4+4
[1] 13771
```{R}

promoter_overlaps <- findOverlaps(pol2_chip, gencode_promoters)
# Note @from (subject) @to (query)


# sometimes it's easier to view these things in data frames
promoter_overlaps_df <- data.frame(pol2_index = promoter_overlaps@from,
                                   promoter_index = promoter_overlaps@to)

overlapping_pol2 <- pol2_chip[promoter_overlaps_df$pol2_index] %>% as.data.frame()
overlapping_promoters <- gencode_promoters[promoter_overlaps_df$promoter_index] %>% as.data.frame()

promoter_overlaps_df$promoter_gene_id <- overlapping_promoters$gene_id
promoter_overlaps_df$promoter_gene_name <- overlapping_promoters$gene_name
promoter_overlaps_df$peak_name <- overlapping_pol2$name


length(unique(promoter_overlaps_df$peak_name))
length(unique(promoter_overlaps_df$promoter_gene_id))
overlapping_pol2
#Let's see how many peaks there were in pol2_chip
length(pol2_chip)
# ok so there are 10,000 or so peaks


# Let's take a look at this object
promoter_overlaps

# It looks like the first peak in Pol 2 overlaps promoter with index 32
# let's verify that
pol2_chip[1]
gencode_promoters[32]

# Peak 1 is pretty broad
width(pol2_chip[1])

# It overlaps 3 promoters. Let's look at another
gencode_promoters[52]

# That's looking at the data by hand
# How do access these values in the overlaps object?
# @ can access a field of an object in R
# In this case @from -- gives up the pol2 peak index
# @to is the promoter index
# If we want to figure out how many promoters each peak overlaps we can use table
table(promoter_overlaps@from)[1:3]

length(table(promoter_overlaps@from))
# Peak 1 overlaps 3 promoters, peak 2 overlaps 2 promoters

table(table(promoter_overlaps@from))

head(promoter_overlaps@from)

table(table(promoter_overlaps@to))
# let's find how many promoter overlaps there were
length(promoter_overlaps@from)
length(unique(promoter_overlaps@from)) 

# this tell us how many peaks overlapped promoters
# There are 13,771 peaks that overlapped promoters
# There are 7,738 peaks that overlap promoters
# Thus roughly each promoter had two peaks overlap


# Let's look at the query file (promoters)

length(promoter_overlaps@to)
# So we see that 13,771 promoters that overlapped peaks?


length(unique(promoter_overlaps@to)) #how many promoters ov polII peaks


# Now we can calculate the percentage of peaks that overlapped promoters
percent_promoter_overlap <- length(unique(promoter_overlaps@from))/length(pol2_chip)
```

Now let's move to working with Gencode and genomic ranges on our data set!
