---
title: "11_R_functions"
author: "JR"
date: "10/20/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("../../../util/intersect_functions.R")
source("../../../util/class_functions.R")
```

Today we will go over a "function" in R. This is a really nice feature of 
defining a function for R that you may run frequently. 

We have used many of R's built in base functions that require parameters too.
For example,table and summary we used before requires a parameter of an object to summarize etc. 
But the cool part is you can make your own functions just like these base R funcitons. 

Now, let's see what it looks like in R. A common function in life is math 
-- it takes inputs and maps it to outputs.

```{r first function}
# This function has two parameters, and we have to let R know (x & y)
# function is inside { }, then we need a 'return' to get the answer

fun <- function(x, y) {
  ans <- x * y
  return(ans)
}
fun(2,-4)

# Note that the object ans doesn't exist in our environment. 
# ans is a local variable that is erased after used. global variables are in our environment.
# It's good to remember the "scope" of vars in a function don't really exist outside the function.
# Finally note a function can only return one thing!
```

When creating a function, you can also add documentation about what that function 
does and the data types and expectationn for the the parameters it takes.
```{r documenting functions}
#' A function to multiply two numbers
#'
#' @description 
#' This function will multiply the input values of X and Y
#' 
#' @param x one number you'd like to multiply
#' y the other number you'd like to multiply
fun <- function(x, y) {
  ans <- x * y
  return(ans)
}

# Let's make a util .R file and put this function in there
# clear environment and load functions -- shows up as something you can just run now
# Note that the documentation is present in the little "note" icon link.
```

Now let's understand "scope" better by creating a more complex function. 

Local variables are those variables that exist only within a certain part of a program 
like a function and are released when the function call ends.

So it's good to remember local variables are mapped to outputs through a function, but 
only exist within a function.

Let's take a look at a more complex function we may want to routinely use.

One common task is to make a list of files to be analyzed and rename them based on information
in the file name itself. Let's look at this function:

```{r base file path to peaks}

# First let's set a file path to where all our peak files are.

broadpeakfilepath <- paste0("/scratch/Shares/rinnclass/CLASS_2022/data/peaks")
# So this is now a global variable that can be called in a function.

# list.files will make a list of all the files in a dir

  fl <- list.files(broadpeakfilepath, 
                   full.names=TRUE)
  # making a list of file names as we did in 08.Rmd Note full file path is included
  # This gives us an index 1-15 where each peak file is in a vector.
  
  fl <- fl[grep("peaks.broadPeak", fl)]
  # making sure we only get files that are .broadPeak using grep
  
```

# sapply function
Let's make a function to find which DBPs have peak files. 
sapply requires a function so let's try it
```{R sapply and regular expression} 

  tf_name <- sapply(fl, function(x){
    y <-  str_extract(x, "([^\\/]+$)")
    unlist(strsplit(y, "_"))[[1]]
    
    # We use a generic function (function(x) and set a new variable y using str_extract
    # then x is subject to str_extract and regex to split all / and words (https://regexr.com/3cr6f)  
    # strplit then splits up the broken up string from str_extract and takes the right side 
        # first character string after the last /
    
    # ?? What would the input be to make this a function ??

  })
```

# write function to import and rename a list of peak files
let's make this into a real function that is re-usable "import_peaks" and name file by DBP

```{R function to rename files by DBP}

# estbalishing the function using "function()"
# establishing the parameter consensus_file_path is needed

import_peaks <- function(consensus_file_path = "/scratch/Shares/rinnclass/CLASS_2022/data/peaks") {
  
  # Setting some variables needed in main part of function (same as above -- peak_files & tf_name)
  peak_files <- list.files(consensus_file_path, full.names = T)
  tf_name <- sapply(peak_files, function(x){
    y <-  str_extract(x, "([^\\/]+$)")
    unlist(strsplit(y, "_"))[[1]]
  })

# Here is the heart of the function that will import each file as GRanges (we can use for overlaps)
  
  peak_list <- c()
  for(i in 1:length(peak_files)) {
    # Import peak files
    peaks <- rtracklayer::import(peak_files[i])
    # Append this GRanges object to the of the list.
    peak_list <- c(peak_list, peaks)
    # Name the list elements by their TF name (we just made above)
    names(peak_list)[length(peak_list)] <- tf_name[i]
  }
  return(peak_list)
}

# ?? Why did nothing come back in the global environment ??

```

# running import peaks function we just made
Let's use import_peaks function but assign the local output to global output! Then recreate "peak_list"

```{R using import_peaks function -- takes a few min to run}

# for didactic purposes we need to create "peak_list" to continue from the functions above 
# This is another good example of how a function only returns local vars and need to assign a global var ( <-)
peak_list <- import_peaks("/scratch/Shares/rinnclass/CLASS_2022/data/peaks")
  
# cool we have a list of Granges nicely named where each entry in the list is a 
# single peak file.

```
 
# Sourcing custom functions -- util folder
Let's add import peaks to util using this header (can take out all the markdown above too):
```{R sourcing custom functions from UTIL}

#' import peak .bed files as a list
#' 
#' @description 
#' this function will take consensus peak files and name them by the DBP
#' and return a list of GRanges peaks for each ChiPseq experiment
#' 
#' @param consensus_file_path the path to consensus peak files
```

# interesect peaks custom function
Cool now we have a function to import peak files as a GRanges.
We can now use this to find overlaps in replicates using another function
"intersect peaks" Let's take a look at our "intersect_peaks" function. 


Here we will use the peak_list we generated above and push it into this function intersect_peaks.
```{r creating intersect_peaks function}

# note the dependence on peak_list which is the output of import_peaks.

intersect_peaks <- function(peak_list) {
  combined_peaks <- peak_list[[1]]
  for(i in 2:length(peak_list)) {
    suppressWarnings(pl_ov <- findOverlaps(combined_peaks, peak_list[[i]]))
    pl1 <- combined_peaks[unique(pl_ov@from)]
    pl2 <- peak_list[[i]][unique(pl_ov@to)]
    suppressWarnings(combined_peaks <- GenomicRanges::reduce(union(pl1, pl2)))
  }
  return(combined_peaks)
}

# Now that we have this function in this environment session we
# could just run it like this:

# overlaps <- intersect_peaks(peak_list)

# convert to data frame
# ov_df <- overlaps %>% as.data.frame()


```

# read_peaks custom function.
The above functions are all part of another function we will get to soon:
'create_consensus_peaks'

This overall function to find overlaps between all 560 
replicates will use the custom functions we just made together:

import_peaks
intersect_peaks 
read_peaks

We did the first two already (in concept) now let's look at read_peaks

```{r read_peaks custom function}

# this function simply filters out the non-cannonical chromosomes.

read_peaks <- function(broad_peak_file, filter_to_canonical_chr = TRUE) {
  dat <- read.table(broad_peak_file, sep = "\t")
  if(filter_to_canonical_chr == TRUE) {
    dat <- dat[dat$V1 %in% c(paste0("chr", 1:22), "chrM", "chrX", "chrY"),]
  }
  gr <- GRanges(seqnames = dat$V1,
                ranges = IRanges(start=dat$V2,end=dat$V3))
  return(gr)
}

# Let's see where the V1-V3 come from in read.table:
test <- read.table("/scratch/Shares/rinnclass/CLASS_2022/data/peaks/CTCF_R1_peaks.broadPeak")

# let's run the function
test2 <- read_peaks("/scratch/Shares/rinnclass/CLASS_2022/data/peaks/CTCF_R1_peaks.broadPeak", filter_to_canonical_chr = TRUE)

# Let's take a look at what we got
table(test2)
# how many peaks are there?
table(table(test2))

# Let's make sure the contigs are gone
test3 <- test2 %>% 
as.data.frame()

# Another way to do the same.
table(test2@seqnames)

# Overall read_peaks filters to normal chromosomes and makes a granges for a given peak file.
# This will be used in a for loop in create_consensus_peaks to filter chromosomes and great a GRanges object -- that will be passed along the for loop to "intersect_peaks" or "findOverlaps"

```

# create_consensus_peaks function
Now let's look at the functions involved in creating consensus peaks.
It is also going to call other custom functions (read_peaks, intersect_peaks )

```{r create_consensus_peaks}
#' intersect replicates into a "consensus peak list" 
#' 
#' @description 
#' this function will take the  union of peak widths across replicates for a given
#' DNA binding protein. the function that will take a list of granges objects and return 
#  one granges object with merged peaks that are in all replicates
#' 
#' @param 
#'  the path to consensus peak files
#' # We're going to iterate over all the files to make it work. 
```

# Review the entire function !
Ok let's go through the code that is a function of functions we went over above.
This is the full function!
```{r reviewing the functions in the functions}

# This is just setting up to name the funciton.
create_consensus_peaks <- function(broadpeakfilepath = "/scratch/Shares/rinnclass/CLASS_2022/data/peaks/") {
  
# For now we can set broadpeakfilepath
  
# broadpeakfilepath <- "/Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/11_consensus_peak_exercise"

# making a list of file paths to the 
  fl <- list.files(broadpeakfilepath, 
                   full.names=TRUE)
  fl <- fl[grep("peaks.broadPeak", fl)]

# getting a DBP name for same index as each file path
  tf_name <- sapply(fl, function(x){
    y <-  str_extract(x, "([^\\/]+$)")
    unlist(strsplit(y, "_"))[[1]]
  })
  
 
  # making sure there is a replicate
  tf_df <- data.frame(table(tf_name)) %>%  # data.frame(table(tf_name))
    filter(Freq > 1)
  unique_tf <- as.character(tf_df$tf_name) # unique_tf
  

# Now a nested for loop (2 for loops) to make GRanges of peak files.
  consensus_peaks <- list()
  for(i in 1:length(unique_tf)) {
    
    # load all the peak files corresponding to this DBP[i] in unique_tf.
    # tf <- unique_tf[1] -- allows us to look at output
    tf <- unique_tf[i]
    print(tf)
    # indexing unique DBP name to file path (e.g., first 8 are CTCF files)
    tf_index <- grep(tf, tf_name)
    tf_files <- fl[tf_index]
    
    # now make a list of GRanges in a peak_list using another for loop
    peak_list <- c()
    for(j in 1:length(tf_files)) {
      # See the read peaks function to know what subfunctions are called.
      peak_list <- c(peak_list, read_peaks(tf_files[j]))
      # same read peaks function and we now have each DBP indexed in tf_files
    }

# filtering chromosomes -- redundant since read peaks does this too -- oh well.
    canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")
    for(i in 1:length(peak_list)) {
      peak_list[[i]] <-peak_list[[i]][which(seqnames(peak_list[[i]]) %in% canonical_chr)]
    }

# Now we use intersect_peaks functino to find overlaps 
    final_peakset <- intersect_peaks(peak_list = peak_list)
    if(length(final_peakset) > 0) {
      final_peakset$name <- paste0(tf, "_", 1:length(final_peakset))
    }
    
    consensus_peaks <- c(consensus_peaks, list(final_peakset))
    names(consensus_peaks)[length(consensus_peaks)] <- tf
  }
  return(consensus_peaks)
}

```


*********************
EXCERCISE
*********************

Make a function for formatting .broadPeak files to UCSC format.

Hint you already wrote all the code in 08 -- compile into function and put in class
functions.R in Util.

