---
title: "12_create_consensus_peaks"
author: "JR"
date: "10/31/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("../../../util/intersect_functions.R")
```

Today we are going to practice running the create_consensus_peaks function. Then import the full dataset of 486 DBPs!

# running consensus peaks -- takes ~3hrs
```{r create_consensus_peaks}

# let's run our new function create consensus peaks -- let's run as local job
# so we first need to source util and the packages needed.

library(tidyverse)
library(GenomicRanges)
source("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/util/intersect_functions.R")

# run consensus peaks
consensus_peaks <- create_consensus_peaks("/scratch/Shares/rinnclass/CLASS_2022/data/peaks")


# export consensus peaks to CORRECT DIRECTORY!
for(i in 1:length(consensus_peaks)) {
  rtracklayer::export(consensus_peaks[[i]], 
                      paste0("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/consensus_peaks/", 
                             names(consensus_peaks)[i], 
                             "_consensus_peaks.bed"))
}

# Nice this only took 30 min locally
# normally is 3hrs in local environment
# long runs are best done as local job outside Rstudio!

# let's take a look
summary(consensus_peaks)
# note CTCF went from 30,000 some peaks to 28,640 that overlap in all 8 replicates

# let's plot the lengths of all the consensus peak files (sapply)
hist(sapply(consensus_peaks, length))

# let's see how wide the peaks are for one example
hist(width(consensus_peaks[[2]]), breaks = 1000)

# let's add a limit to the x-axis so we cna see this better
hist(width(consensus_peaks[[2]]), breaks = 1000, xlim = c(0, 300))

# phew the peaks are mostly bigger than 100 bp.

```


# filtering consensus peaks to at least 250 peaks

```{r filtering consensus peaks > 250}

num_peaks_threshold <- 250
filtered_consensus_peaks <- consensus_peaks[sapply(consensus_peaks, length) > num_peaks_threshold]
# Now we have a shorter list of GRanges!
summary(filtered_consensus_peaks)
table(summary(filtered_consensus_peaks))
# Cool we see that they all have more than 250 peaks 


# Now let's export the filtered consensus peaks
for(i in 1:length(filtered_consensus_peaks)) {
  rtracklayer::export(consensus_peaks[[i]], 
                      paste0("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/filtered_consensus_peaks/", 
                             names(consensus_peaks)[i], 
                             "_filtrered_consensus_peaks.bed"))
}

# let's see what we lost and record it.
# setdiff is a function to determine what is NOT in another list

# order matters want the biggest list first to find what is not in it.
lost_dbp <- setdiff(names(consensus_peaks), names(filtered_consensus_peaks))
lost_dbp

# let's see how order matters
lost_dbp2 <- setdiff(names(filtered_consensus_peaks), names(consensus_peaks))

# we can also do this by indexing too

lost_dbp <- names(consensus_peaks)[!(names(consensus_peaks) %in% 
                                       names(filtered_consensus_peaks))]

# let's write this out for our results of which had < 250 peaks
write.table(lost_dbp, "lost_dbps.csv")

```

# START POINT: reload in consensu peaks
Moving forward we will often start by loading the consensus peaks 
Below is the code we will use to do that

```{r START POINT reimporting filtered consensus peaks}

# making a file list
filtered_consensus_peaks_files <- list.files("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/filtered_consensus_peaks", 
                                             pattern = "*.bed",
                                             full.names = TRUE)

# like a for loop we can use lapply to recursively load each file in the "peak_files"
filtered_consensus_peaks <- lapply(filtered_consensus_peaks_files, rtracklayer::import)
# what is missing here??? Let's look at the object ...

# This is adding the names to each GRange in the list -- names are very important :) 
names(filtered_consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2022/data/consensus_peaks/|filtered_consensus_peaks.bed", "", filtered_consensus_peaks_files)

```

# Creating genome annotation files
Great, now we have a bunch of data -- let's see where the peaks land in different annotations of the genome.

First let's make GRanges of mRNA and lncRNA promoter regions (save somewhere handy in your local folder)

```{r defininig promoters for specific gene lists}

# first load in the gencode annotation file
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/genomes/gencode.v32.annotation.gtf")

# lncRNA and mRNA promoters (half annotations are pseudogenes and other types)
lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))

# export to save for later
rtracklayer::export(lncrna_mrna_promoters, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncRNA_mrna_promoters.gtf")

# let's do a test reimport as we will do this often too.
lncrna_mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncRNA_mrna_promoters.gtf")

# note we can add names!
names(lncrna_mrna_promoters) <- lncrna_mrna_promoters$gene_id

# creating lncRNAs promoter features
lncrna_promoters <- get_promoter_regions(gencode_gr, biotype = "lncRNA")
rtracklayer::export(lncrna_promoters, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_promoters.gtf")

# let's reimport and add names
lncrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_promoters.gtf")



# finally mRNA promoters

mrna_promoters <- get_promoter_regions(gencode_gr, biotype = "protein_coding")
rtracklayer::export(mrna_promoters, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/mrna_promoters.gtf") 

# let's reimport and add names

mrna_promoters <- rtracklayer::import( "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/mrna_promoters.gtf") 


mrna_promoters_df <- mrna_promoters %>%
  as.data.frame()


```

# Making gene body annotation files.
Now let's do the same for gene bodies. 
This will primarily be done with clever indexing 

```{r making gene bodie annotation files}

# lncRNA and mRNA gene bodies
lncrna_mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                     gencode_gr$gene_type %in% c("lncRNA", "protein_coding")]
# exporting
rtracklayer::export(lncrna_mrna_genebody, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_mrna_genebody.gtf")

# lncRNA genebodies
lncrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                gencode_gr$gene_type %in% c("lncRNA")]

# exporting
rtracklayer::export(lncrna_genebody, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_genebody.gtf")

# mrna gene bodies
mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                              gencode_gr$gene_type %in% c("protein_coding")]

# exporting
rtracklayer::export(mrna_genebody, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/mrna_genebody.gtf")


```


################
ExceRcise
################

Pick a protein in the larger data set and see how many peaks overlap mRNAs and lncRNAs.
Hint: findOverlaps

```{R exeRcise hint}

#protein_x <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2022/data/consensus_peaks/AFF4.bed")

#ov_x <- findOverlaps(protein_x, lncrna_mrna_promoters)

#length(ov_x)
#length(protein_x)
#length(lncrna_mrna_promoters)


#ov_x_gene_bodies <- findOverlaps(protein_x, lncrna_mrna_genebody)

#length(ov_x_gene_bodies)
#length(protein_x)
#length(lncrna_mrna_genebody)

```


Now, use your function to convert the .bed files created above to check them in the USCS browser.
Pick 2 similar proteins and look at the peak files and consensus peaks files in UCSC browser.

Do the peaks seem to overlap?

