---
title: "12_create_consensus_peaks"
author: "JR"
date: "10/31/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("util/intersect_functions.R")
```



Today we are going to practice running the create_consensus_peaks function. Then import the full dataset of 486 DBPs!

First let's run 
```{r}
consensus_peaks <- create_consensus_peaks("/Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/11_consensus_peak_exercise")



for(i in 1:length(consensus_peaks)) {
  rtracklayer::export(consensus_peaks[[i]], 
                      paste0("/scratch/Shares/rinnclass/consensus_peaks/", 
                             names(consensus_peaks)[i], 
                             "_consensus_peaks.bed"))
}
```


Let's filter on number of peaks 
```{r}
num_peaks_threshold <- 250
filtered_consensus_peaks <- consensus_peaks[sapply(consensus_peaks, length) > num_peaks_threshold]



summary(filtered_consensus_peaks)
# Cool we see that CTCF and DLX6 are more than 10,000 peaks.
# Last year we filtered to those with 250 or peaks -- turned out to be a reasonable quartile.

# Now let's export the consensus peaks

for(i in 1:length(consensus_peaks)) {
  rtracklayer::export(consensus_peaks[[i]], 
                      paste0("/Shares/rinn_class/data/CLASS_2022/data/filtered_consensus_peaks/", 
                             names(consensus_peaks)[i], 
                             "_consensus_peaks.bed"))
}



```

Let's go ahead and import all the DBPs we are studying 483!
The files are located at:
/scratch/Shares/rinnclass/CLASS_2022/data/consensus_peaks
```{r}

consensus_peaks_files <- list.files("/scratch/Shares/rinnclass/CLASS_2022/data/consensus_peaks", 
                                             pattern = "*.bed",
                                             full.names = TRUE)

# like a for loop we can use lapply to recursively load each file in the "peak_files"
consensus_peaks <- lapply(consensus_peaks_files, rtracklayer::import)
# what is missing here??? Let's look at the object ...

# This is adding the names to each GRange in the list -- names are very important :) 
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2022/data/consensus_peaks/|_filtered_consensus_peaks.bed", "", consensus_peaks_files)

```


Great, now we have a bunch of data -- let's see where the peaks land in different annotations of the genome.

First let's make GRanges of mRNA and lncRNA promoter regions (save somewhere handy in your local folder)

1) Promoter Regions
```{r}
gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))
rtracklayer::export(lncrna_mrna_promoters, "/Shares/rinn_class/data/CLASS_2022/data/genome_features/lncrna_mrna_promoters.gtf")

lncrna_promoters <- get_promoter_regions(gencode_gr, biotype = "lncRNA")
rtracklayer::export(lncrna_promoters, "/Shares/rinn_class/data/CLASS_2022/data/genome_features/lncrna_promoters.gtf")

mrna_promoters <- get_promoter_regions(gencode_gr, biotype = "protein_coding")
rtracklayer::export(mrna_promoters, "/Shares/rinn_class/data/CLASS_2022/data/genome_features/mrna_promoters.gtf")
```


Same for gene bodies.

2) Gene Bodies

```{r}
lncrna_mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                     gencode_gr$gene_type %in% c("lncRNA", "protein_coding")]
rtracklayer::export(lncrna_mrna_genebody, "/Shares/rinn_class/data/CLASS_2022/data/genome_features/lncrna_mrna_genebody.gtf")

lncrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                gencode_gr$gene_type %in% c("lncRNA")]
rtracklayer::export(lncrna_genebody, "/Shares/rinn_class/data/CLASS_2022/data/genome_features/lncrna_genebody.gtf")

mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                              gencode_gr$gene_type %in% c("protein_coding")]
rtracklayer::export(mrna_genebody, "/Shares/rinn_class/data/CLASS_2022/data/genome_features/mrna_genebody.gtf")
```

Excercise 1:

Pick a protein in the larger data set and see how many peaks overlap mRNAs and lncRNAs.
Hint: findOverlaps

```{r}


#protein_x <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2022/data/consensus_peaks/AFF4.bed")

#ov_x <- findOverlaps(protein_x, lncrna_mrna_promoters)

#length(ov_x)
#length(protein_x)
#length(lncrna_mrna_promoters)


#ov_x_gene_bodies <- findOverlaps(protein_x, lncrna_mrna_genebody)

#length(ov_x_gene_bodies)
#length(protein_x)
#length(lncrna_mrna_genebody)

```



Now, use you function to convert the .bed files created above to check them in the USCS browser.
Pick 2 proteins and look at the peak files and consensus peaks files in UCSC browser.

Do the peaks seem to overlap?

```{r}






```


