---
title: "12_create_consensus_peaks"
author: "JR"
date: "10/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("../../util/intersect_functions.R")
```



Today we are going to practice running the create_consensus_peaks function. Then import the full dataset of 486 DBPs!

First let's run 
```{r}
consensus_peaks <- create_consensus_peaks("/Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/11_consensus_peak_exercise")
for(i in 1:length(consensus_peaks)) {
  rtracklayer::export(consensus_peaks[[i]], 
                      paste0("/scratch/Shares/rinnclass/consensus_peaks/", 
                             names(consensus_peaks)[i], 
                             "_consensus_peaks.bed"))
}
```

That was a good exercise for a few files and the same process already took place for all 486 DBPs!
The files are located at:

/scratch/Shares/rinnclass/CLASS_2022/data/consensus_peaks

Let's use import peaks to make a GRange obnject.
```{r}
import
```


Let's filter 
```{r}
num_peaks_threshold <- 10000
filtered_consensus_peaks <- consensus_peaks[sapply(consensus_peaks, length) > num_peaks_threshold]
#Export consensus peaks
for(i in 1:length(filtered_consensus_peaks)) {
  rtracklayer::export(filtered_consensus_peaks[[i]], 
                      paste0("/scratch/Shares/rinnclass/YOUR_FOLDER/X/results/chipseq/filtered_consensus/", 
                             names(filtered_consensus_peaks)[i], 
                             "_filtered_consensus_peaks.bed"))
}

summary(filtered_consensus_peaks)
# Cool we see that CTCF and DLX6 are more than 10,000 peaks.
# Last year we filtered to those with 250 or peaks -- turned out to be a reasonable quartile.

```

Let's go ahead and import all the DBPs we are studying 486!
```{r}

```










Exporting Genomic features that we will use frequentyl so we can use rtracklayer::import and save a ton of time :)

These you will want to save directly to your results folder since they will be used for all analyses and not just chipseq (where we exported consensus and filtered consensus peaks)

1) Promoter Regions
```{r}
gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")
lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))
rtracklayer::export(lncrna_mrna_promoters, "/scratch/Shares/rinnclass/YOUR_FOLDER/analysis/--- 00 consensus peaks -- /lncrna_mrna_promoters.gtf")
lncrna_promoters <- get_promoter_regions(gencode_gr, biotype = "lncRNA")
rtracklayer::export(lncrna_promoters, "/scratch/Shares/rinnclass/YOUR_FOLDER/analysis/--- 00 consensus peaks -- /lncrna_promoters.gtf")
mrna_promoters <- get_promoter_regions(gencode_gr, biotype = "protein_coding")
rtracklayer::export(mrna_promoters, "/scratch/Shares/rinnclass/YOUR_FOLDER/analysis/--- 00 consensus peaks -- /mrna_promoters.gtf")
```


2) Gene Bodies

```{r}
lncrna_mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                     gencode_gr$gene_type %in% c("lncRNA", "protein_coding")]
rtracklayer::export(lncrna_mrna_genebody, "/scratch/Shares/rinnclass/YOUR_FOLDER/analysis/--- 00 consensus peaks -- /lncrna_mrna_genebody.gtf")
lncrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                gencode_gr$gene_type %in% c("lncRNA")]
rtracklayer::export(lncrna_genebody, "/scratch/Shares/rinnclass/YOUR_FOLDER/analysis/--- 00 consensus peaks -- /lncrna_genebody.gtf")
mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                              gencode_gr$gene_type %in% c("protein_coding")]
rtracklayer::export(mrna_genebody, "/scratch/Shares/rinnclass/YOUR_FOLDER/analysis/--- 00 consensus peaks -- /mrna_genebody.gtf")
```


