---
title: "12_create_consensus_peaks"
author: "JR"
date: "10/31/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
# source("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/util")
source("/scratch/Shares/rinnclass/CLASS_2022/<your_folder/CLASS_2022/util/intersect_functions.R")
```

Today we are going to practice running the create_consensus_peaks function. Then import the full dataset of 483 DBPs!

# running consensus peaks -- 30 min or so
# We are already running create_consensus_peaks form 11_consensus_peaks.RMD 
# So we can skip this but useful starting point.
```{r create_consensus_peaks -- run as local job}

# let's run our new function create consensus peaks -- let's run as local job
# so we first need to source util and the packages needed.

# now we need to make an Rscript and put in working dir (11)

# library(tidyverse)
# library(GenomicRanges)
# source("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/util/intersect_functions.R")
# source("/scratch/Shares/rinnclass/CLASS_2022/<your_folder>/CLASS_2022/util/intersect_functions.R")

# run consensus peaks
# consensus_peaks <- create_consensus_peaks("/scratch/Shares/rinnclass/CLASS_2022/data/peaks")
```

# export consensus peaks to CORRECT DIRECTORY!

```{R exporting consensus peaks to 11 directory}

# for(i in 1:length(consensus_peaks)) {
#  rtracklayer::export(consensus_peaks[[i]], 
#                     paste0("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/consensus_peaks/", 
#                             names(consensus_peaks)[i], 
#                             "_consensus_peaks.bed"))
# }

# Run local job by setting working dir to 11 - ref .R script

```

Great, now we have a bunch of data -- let's see where the peaks land in different annotations of the genome.
First let's make GRanges of mRNA and lncRNA promoter regions (save somewhere handy in your local folder)

# Making annotaiton files while consensus peaks runs:
# Creating genome annotation files

```{r reading in gencode annotations}

# first load in the gencode annotation file
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/genomes/gencode.v32.annotation.gtf")

```

## START POINT CONSENSUS PEAK SUMMARY ## 

# Consensus Peak Properties

```{R consensus_peak properties}

# make a list of file paths for consensus peaks
consensus_fl <- list.files("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/consensus_peaks", full.names = T)

# using lapply to import all the files in consensus_fl
consensus_peaks <- lapply(consensus_fl, rtracklayer::import)

# adding names from the file path
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/consensus_peaks/|_consensus_peaks.bed","", consensus_fl)


# let's take a look
summary(consensus_peaks)
# note CTCF went from 30,000 some peaks to 28,640 that overlap in all 8 replicates

num_peaks <- sapply(consensus_peaks, length)

# let's plot the lengths of all the consensus peak files (sapply)
hist(num_peaks)

# Let's break this down more and look into peaks near 0

hist(num_peaks, breaks = 1000, xlim = c(0, 1000))
summary(num_peaks)

max(num_peaks)
min(num_peaks)
num_peaks[372]

which.max(num_peaks)

consensus_peaks[which.max(num_peaks)]

# let's see how wide the peaks are for one example
hist(width(consensus_peaks[[37]]), breaks = 1000)

# let's add a limit to the x-axis so we cna see this better
hist(width(consensus_peaks[[37]]), breaks = 1000, xlim = c(0, 300))
# we see that there aren't any peaks shorter than 175 
# phew the peaks are mostly bigger than 100 bp.

```


# filtering consensus peaks to at least 250 peaks
# Do we want a different number??

```{r filtering consensus peaks > 250}

num_peaks_threshold <- 250

num_peaks <- sapply(consensus_peaks, length)

filtered_consensus_peaks <- consensus_peaks[num_peaks > num_peaks_threshold]
# Now we have a shorter list of GRanges!

summary(filtered_consensus_peaks)
table(summary(filtered_consensus_peaks))
# Cool we see that they all have more than 250 peaks 


# Now let's export the filtered consensus peaks
for(i in 1:length(filtered_consensus_peaks)) {
  rtracklayer::export(filtered_consensus_peaks[[i]], 
                      paste0("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/filtered_consensus_peaks/", 
                             names(consensus_peaks)[i], 
                             "_filtered_consensus_peaks.bed"))
}

# let's see what we lost and record it.
# setdiff is a function to determine what is NOT in another list

# order matters want the biggest list first to find what is not in it.
lost_dbp <- setdiff(names(consensus_peaks), names(filtered_consensus_peaks))
?setdiff

ASYMETRIC <- setdiff(names(filtered_consensus_peaks), names(consensus_peaks))

lost_dbp

lost_dbp <- names(consensus_peaks)[!(names(consensus_peaks) %in% 
                                       names(filtered_consensus_peaks))]
lost_dbp

# let's write this out for our results of which had < 250 peaks
write.table(lost_dbp, "lost_dbps.csv")

```

Moving forward we will often start by loading the consensus peaks 
Below is the code we will use to do that

```{r START POINT reimporting filtered consensus peaks}

# making a file list
filtered_consensus_peaks_files <- list.files("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/filtered_consensus_peaks", 
                                             pattern = "*.bed",
                                             full.names = TRUE)

# like a for loop we can use lapply to recursively load each file in the "peak_files"
filtered_consensus_peaks <- lapply(filtered_consensus_peaks_files, rtracklayer::import)
# what is missing here??? Let's look at the object ...

# This is adding the names to each GRange in the list -- names are very important :) 
names(filtered_consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2022/data/consensus_peaks/|filtered_consensus_peaks.bed", "", filtered_consensus_peaks_files)

```


## START POINT CLASS 3/9

First: please set working directory to 11.
Second: create a "gene_annotation_files" directory

Third: lets load in the needed packages and gencode annotations

```{R lncRNA and mRNA promoters}

# First we need to source "get_promoter_regions" and other functions from your util :
source("/scratch/Shares/rinnclass/CLASS_2022/<your_folder/CLASS_2022/util/intersect_functions.R")
# source("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/util/intersect_functions.R")

library(tidyverse)
library(GenomicRanges)

# loading Gencode annotaitons to parse out data into files for future use:
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/genomes/gencode.v32.annotation.gtf")

```

# Creating promoter Ranges from Gencode
Let's start making annotation files for promoters and gene bodies
We will use these files moving forward.
Bonus: we won't have to load gencode annotations and wait as much :)

# NOTE: handy copy tab to present code of GRanges !!
```{R exporting promoter annotations from gencode}
# lncRNA and mRNA promoters (half annotations are pseudogenes and other types)
lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))
gencode_gr@biotype

# Look at what we get in the environment.

# export to save for later
rtracklayer::export(lncrna_mrna_promoters, "file_path/lncRNA_mrna_promoters.gtf")

# "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncRNA_mrna_promoters.gtf")

# let's do a test reimport as we will do this often too.
lncrna_mrna_promoters <- rtracklayer::import("file_path/lncRNA_mrna_promoters.gtf")

# lncrna_mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncRNA_mrna_promoters.gtf")

# Adding the names of the genes to the promoter.
# First let's look into the Granges more to see where gene_id is :
lncrna_mrna_promoters@elementMetadata@listData[["gene_id"]]

# Now add names and see what we get
names(lncrna_mrna_promoters) <- lncrna_mrna_promoters$gene_id

names(lncrna_mrna_promoters)
# Cool now the names are there 

```

# Now let's make anntoation files for lncRNA and mRNA promoters seperately

```{R making lncRNA and mRNA promoter files seperately}

# creating lncRNAs promoter features
lncrna_promoters <- get_promoter_regions(gencode_gr, biotype = "lncRNA")

# Take a look in environment 
# NOTE: copy tab for getting code to index:
lncrna_promoters@elementMetadata@listData[["gene_name"]]
# this gives us all the gene names: Note index 1 is a microRNA :) 

# exporting
rtracklayer::export(lncrna_promoters, "file_path/lncrna_promoters.gtf")

# rtracklayer::export(lncrna_promoters, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_promoters.gtf")

# let's reimport and add names
lncrna_promoters <- rtracklayer::import("file_path/lncrna_promoters.gtf")

#lncrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_promoters.gtf")


# finally mRNA promoters
mrna_promoters <- get_promoter_regions(gencode_gr, biotype = "protein_coding")

# Take a look from index tab in env.
mrna_promoters@elementMetadata@listData[["gene_name"]]


# exporting
rtracklayer::export(mrna_promoters, "/file_path/mrna_promoters.gtf") 

# rtracklayer::export(mrna_promoters, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/mrna_promoters.gtf") 

# let's reimport 
mrna_promoters <- rtracklayer::import( "file_path/mrna_promoters.gtf") 

# mrna_promoters <- rtracklayer::import( "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/mrna_promoters.gtf") 


# Let's look more closely with a data frame.
mrna_promoters_df <- mrna_promoters %>%
  as.data.frame()

# Nice and easy to read, shows how easy to use GRanges then convert to DF.

```

# Making gene body annotation files.
Now let's do the same for gene bodies. 
This will primarily be done with clever indexing 

```{r making gene body annotation files}

# Let's first look at what we are indexing:
encode_gr$gene_type

# so from this list we want to grab all the lncRNA genes and "protein_coding" genes


# lncRNA and mRNA gene bodies
lncrna_mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                     gencode_gr$gene_type %in% c("lncRNA", "protein_coding")]

# Let's double check by checking widths in GRanges
table(lncrna_mrna_genebody@ranges@width)


# exporting
rtracklayer::export(lncrna_mrna_genebody, "file_path/lncrna_mrna_genebody.gtf")

# rtracklayer::export(lncrna_mrna_genebody,"/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_mrna_genebody.gtf")


# lncRNA genebodies
lncrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                gencode_gr$gene_type %in% c("lncRNA")]

# exporting
rtracklayer::export(lncrna_genebody, "file_path/lncrna_genebody.gtf")

# rtracklayer::export(lncrna_genebody, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_genebody.gtf")

# mrna gene bodies
mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                              gencode_gr$gene_type %in% c("protein_coding")]

# exporting
# rtracklayer::export(mrna_genebody, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/mrna_genebody.gtf")

rtracklayer::export(mrna_genebody, "file_path/mrna_genebody.gtf")


```

###############
ExceRcise
################

Pick a protein in the larger data set and see how many peaks overlap mRNAs and lncRNAs.
Hint: findOverlaps

```{R exeRcise hint}

#protein_x <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2022/data/consensus_peaks/AFF4.bed")

#ov_x <- findOverlaps(protein_x, lncrna_mrna_promoters)

#length(ov_x)
#length(protein_x)
#length(lncrna_mrna_promoters)


#ov_x_gene_bodies <- findOverlaps(protein_x, lncrna_mrna_genebody)

#length(ov_x_gene_bodies)
#length(protein_x)
#length(lncrna_mrna_genebody)

```

###############
ExceRcise 2
################

Now, use your function to convert the .bed files created above to check them in the USCS browser.
Pick 2 similar proteins and look at the peak files and consensus peaks files in UCSC browser.

Do the peaks seem to overlap?

