---
title: "13_ChIP_peak_features"
author: "JR"
date: "11/1/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("util/intersect_functions.R")
```

Today we are going to explore where in the genome the 483(460) DBPs bind.

From now on we will import all the data we made in "12_creating_consensus_peaks"
Let's start importing all the files we need:
```{r}
# starting with filtered_consensus_peaks

filtered_consensus_peaks_files <- list.files("/Shares/rinn_class/data/CLASS_2022/data/filtered_consensus_peaks", 
                                             pattern = "*.bed",
                                             full.names = TRUE)
filtered_consensus_peaks <- lapply(filtered_consensus_peaks_files, rtracklayer::import)

names(filtered_consensus_peaks) <- gsub("/Shares/rinn_class/data/CLASS_2022/data/filtered_consensus_peaks/|_consensus_peaks.bed", "", filtered_consensus_peaks_files)

names(filtered_consensus_peaks) <- gsub("/Shares/rinn_class/data/CLASS_2022/data/filtered_consensus_peaks/|_consensus_peaks.bed|.bed", "", filtered_consensus_peaks_files)
```

Now we will import the GRanges of genome annotations:
```{R}

lncrna_mrna_promoters <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2022/data/genome_features/lncrna_mrna_promoters.gtf")

lncrna_mrna_genebody <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2022/data/genome_features/lncrna_mrna_genebody.gtf")

# Let's also make a list of gene_ids. Note the indexing.
lncrna_gene_ids <- lncrna_mrna_genebody$gene_id[lncrna_mrna_genebody$gene_type == "lncRNA"]

mrna_gene_ids <- lncrna_mrna_genebody$gene_id[lncrna_mrna_genebody$gene_type == "protein_coding"]
```


Now we are going to start to make a really useful data-frame (df).
For each property/overlap we will make a new column. 
```{r}
# Let's start with loading in the number of peaks each DBP has -- using length.

num_peaks_df <- data.frame("dbp" = names(filtered_consensus_peaks),
                           "num_peaks" = sapply(filtered_consensus_peaks, length))


# Now let's get the total amount of the genome covered by all the peaks for a given DBP.

num_peaks_df$total_peak_length <- sapply(filtered_consensus_peaks, function(x) sum(width(x)))

# Let's take a look at what we have so far (which DBP, has most peaks? most genome coverage?)
```

Great we are starting to build an informative DF to incorporate our results.
Now let's start looking at how many peaks overlap the genome feature files we made.
```{r}

# First let's get a count for each DBP that overlaps any promoter.
## This takes a minute to run :) Let's look count_peaks_per_feature function in meantime

promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, type = "counts")

# Take a look at "count_peaks_per_feature" maybe look at in debugging mode?

```

Above we made the object "promoter_peak_counts" now we can extract information from it
```{R}

# First let's get a count of how many promter overlapped each DBP.

num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)

# Now let's break these promoters into two groups "lncrna" and "mrna"
# We will use the gene_id objects we made above to index and separate them.

num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])

num_peaks_df$peaks_overlapping_mrna_promoters <- 
  rowSums(promoter_peak_counts[,mrna_gene_ids])

# Let's take a look.

```

Now let's do the same thing for gene-bodies
First we will make a gene-body object like we did for promoters above.
```{r}

# Finding overlaps with gene_bodies (will take a few minutes again)

genebody_peak_counts <- count_peaks_per_feature(lncrna_mrna_genebody, 
                                                filtered_consensus_peaks, 
                                                type = "counts")


# Now let's extract the overlaps the same way we did for promoters above

# All gene bodies
num_peaks_df$peaks_overlapping_genebody <- 
  rowSums(genebody_peak_counts)

# lncRNA gene bodies 
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])

# mRNA gene bodies
num_peaks_df$peaks_overlapping_mrna_genebody <- 
  rowSums(genebody_peak_counts[,mrna_gene_ids])

# Now let's take a look.

```

We have a large group of different DBPs so let's import some annotaitons of which
ones are transcription factors etc. A paper in Cell is one of the better annotations --
however they are never perfect :)

Human TF annotations:
https://www.cell.com/cms/10.1016/j.cell.2018.01.029/attachment/ede37821-fd6f-41b7-9a0e-9d5410855ae6/mmc2.xlsx

Let's download the file using R (we could also use curl/wget in BASH)
```{r}

url <- "https://www.cell.com/cms/10.1016/j.cell.2018.01.029/attachment/ede37821-fd6f-41b7-9a0e-9d5410855ae6/mmc2.xlsx"

destination_for_url <- "/Shares/rinn_class/data/CLASS_2022/data/TF_annotations.xlsx"

# to download we can use download.file

download.file(url, destination_for_url)

# The above is a handy bit of code for downloading data into R directly,but BASH works too

# Download the file locally and let's look it over to see what we might want.
```


Now that we have the file we need to read it in. Since it's excel file we will use
a new 'read' function (we have used 'read.table' & 'read.csv' etc).

To this end we will use readX1 'read_excel' function.
```{R}

human_tfs <- readxl::read_excel("/Shares/rinn_class/data/CLASS_2022/data/TF_annotations.xlsx",
                                sheet = 2, skip = 1)
names(human_tfs)[4] <- "is_tf"

length(which(tolower(num_peaks_df$dbp) %in% tolower(human_tfs$Name)))


human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(num_peaks_df$dbp), 1:4]
names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)

# Let's check how many NAs -- we should have some missing values.

write_csv(num_peaks_df, "results/num_peaks_df.csv")
```


```{r}
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, 
                                               type = "occurrence")
# Output to promoter_peak_occurecne_matrix
write.table(promoter_peak_occurence, "results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")
# Now we want to make into a data frame using the promoter annotations as rows and attributes as columns.
# We will use lncrna_mrna_promoters to index "all promoters"
# First make sure promoter_peak_occurence and lncrna_mrna_promoters are in the same order
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))



# We are going to use the promoter peak occurence matrix above to essentially
# recreate a working version of num_peaks_df. However we will now organize it more
# this is a good example of how to set up a bunch of columns in dataframe. using the
# data.frame() fucntion.
# essentially we will index values from the objects we created and make a .CSV 
# to keep adding onto in future classes.

peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "3kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))
# This is the CSV file we will start building upon adding columns of properties as we analyze them
# The output file name will change based on what is added later, but the "peak_occurence_df" will be used throughout.
write_csv(peak_occurence_df, "results/peak_occurence_dataframe.csv")
```

Excercise

Broom environment
create a new directory in your analysis directory on your branch
hint: make folders for each type of analysis (00 consensus peaks / 01 overlapps num_peaks_df)
Load all the files needed to make the plots in 10_


Since there are roughly similar lncRNA and mRNA promoters. Figure out which DBP seems to bind
more lncRNA promoters than mRNA promoters 
