---
title: "15_clustering"
author: "JR"
date: "10/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggdendro)
library(GenomicRanges)
source("../../../util/intersect_functions.R")
source("../../../util/plotting_functions.R")
source("../../../util/_setup.R")
```

Today we are going to use hierarchical clustering to identify DBPs with similar binding patterns.

In brief, this involves calculating the distance (euclidian or spearman) between all pairwise 
comparisons of DBP binding profiles. We have made a matrix of 1 and 0 if a DBP is bound at a 
given genome annotation feature (e.g., promoter, gene-body). This is a binary clustering and 
if we cluster a continuous variable (e.g. RNAseq expression levels) then we may want to make 
other considerations for clustering.

We can use those vectors of 1 and 0 to determine how similar the binding profiles are.

Ok let's get started by grabbing the peak_occurrence_matrix from 14_2_our_data_in_ranges.

```{r}

#let's look at the promoter peak occurence matrix (takes a min to load)

promoter_peak_occurence_matrix <- read.table("/Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/13_ChIP_peak_features/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

# Ok we see this is a very large matrix of 1 and 0 if a DBP is bound to a promoter.
# We can use these 36,000+ vectors of genes with 483 values of bound (1) or unbound (0)


# let's filter to DBPs with at least 250 binding events or the matrix is a bit sparse.
peak_occurence_matrix <- promoter_peak_occurence_matrix[rowSums(promoter_peak_occurence_matrix) > 250, ]

```

Let's get started by finding the 'distance' between each promtoer vector.
We will make a distance matrix of 451 x451 and then use that for clustering.
```{r}

# First let's take a look at the matrix

peak_occurence_matrix[1:2,1:40]

# now we can calculate the distance of DBP (we could also do this for promoters)
# let's do a small test first as this can take a while to calculate 451x451!

test <- dist(peak_occurence_matrix[1:2,], method = "euclidean")
test

# There are a number of measures to calculate distance between two vectors and we can check those out in the documentation.
?dist


# let's see how this changes if we use binary as the distance method
# since this is a binary vector we will use a binary distance metric

test2 <- dist(peak_occurence_matrix[1:2,], method = "binary")
test2


# Now let's create distance matrix for each pairwise vector comparison for the whole matrix now.

peak_occurence_dist <- dist(peak_occurence_matrix, method = "binary")

```

Now that we have that, we can do hierarchical clustering. In brief, this will cluster 
vectors that have small distances (high correlation), and will iterate that process until
everything is in the same cluster.


```{R}
# let's take a look at hclust

?hclust


# Alright we are ready to cluster everything!

bin_hier <- hclust(peak_occurence_dist, method = "complete")

# we now need to plot the information in bin_hier
plot(bin_hier)

# To make this bigger, we can plot it as a pdf
pdf("figures/dbp_hclust_dendro.pdf", height = 12, width = 70)
plot(bin_hier)
dev.off()

bin_hier$labels[bin_hier$order]
```

Nice so we now have object "bin_hier" that will contain all the cross correlation values across all the samples. 

Now we get to PLOT !! We will use ggdenro package that will plot the branch lengths that indicate how similar two samples are -- the longer the branch the more different they are -- closer more similar.

Here is an alternative way to plot it:

```{r}


# First we are going to order the samples by the correlation matrix we just made.

# You can also use ggdendro to plot with ggplot syntax -- which looks nice but for this many 
# DBPs, it's a little bit easier to visualize with base R.

# Now we can plot
# ggdendro::ggdendrogram(bin_hier, rotate = FALSE,  size = 3, 
#                        theme_dendro = TRUE) +
#   coord_flip() +
#   # scale_y_continuous() +
#   # scale_x_continuous(position = "top") +
#   scale_x_continuous(breaks = seq_along(bin_hier$labels[bin_hier$order]),
#             labels = bin_hier$labels[bin_hier$order], position = "top",
#             expand = c(0,0)) +
#   theme(axis.text.x = element_text(angle = 90, hjust  = 1)) + 
#   theme(axis.text.y = element_text(angle = 0,hjust = 1)) +
#   scale_y_reverse(expand = c(0.01, 0)) +
#   theme(
#     plot.background = element_blank(),
#     panel.grid.major = element_blank(),
#     panel.grid.minor = element_blank(),
#     panel.border = element_blank()
#   )

# Go over the plotting aspects as we discussed in 11_intro_plotting_R.Rmd
# Especially thje scale_x_continious part -- we saw that in the 11_ class 

# So this is very difficult to see since we have SO much data :)
# one thing you can see is "phospho S5" and "phospho S2".
# These are modifications on Pol II that indicate two different states:
# S2 = distal or elongating Pol II so less on promoter more in genebody
# S5 = Indicative of Pol II at promoter proximilar regions.
# Note how these don't cluster anywhere near each other -- that is good news!


# Let's save this as a figure -- we want to keep organized so we made a ChIPseq
# folder and figures inside that folder. Let's put the figure there:

# ggsave("chipseq/figures/all_hclust_binary_dist.pdf", height = 44, width = 5.0)

# Nice, this will make it much easier to read :)
# use file transfer to bring to desktop and take a look!
# This is looking biologically relevant! Notice Pol II phospho 2 is
# similar to all other compents of Pol II -- that is exactly what we expect.
# also Pol II S5 is associated with H3K27ac which is a mark of active promoters
# again what we would expect of promoter proximal Pol II.
# MED DBPs all near each other too!
# Sweet -- with a simple spot check we can see all of this data is making sense!
```

Here's some code to cluster mRNA promoters and lncRNA promoters separately. They end up looking pretty similar so we'll skip it for today.

```{r}

# # Let's start with lncRNA promoters -- why not?
# 
# # We first need to load the .GTF we made for lncRNAs and mRNAs.
# 
# lncrna_promoters <- rtracklayer::import("results/lncrna_promoters.gtf")
# mrna_promoters <- rtracklayer::import("results/mrna_promoters.gtf")
# 
# # Cool now we can do the same as above:
# 
# lncrna_peak_occurence <- peak_occurence_matrix[,lncrna_promoters$gene_id]
# 
# bin_hier_lncrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))
# 
# ggdendro::ggdendrogram(bin_hier_lncrna, rotate = TRUE,  size = 3)
# 
# ggsave("chipseq/figures/lncrna_hclust_binary_dist.pdf", height = 44, width = 6)
# 
# # Cool we can see the Pol II S2 and S5 diff right away again!
# 
# 
# # Now for mRNA
# 
# mrna_peak_occurence <- peak_occurence_matrix[,mrna_promoters$gene_id]
# 
# bin_hier_mrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))
# 
# ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)
# 
# ggsave("chipseq/figures/mrna_hclust_binary_dist.pdf", height = 44, width = 6)


# Do a file transfer and open up lncRNA and mRNA clusters side by side:
# They are almost identical !! That is another good sign.

# Nice I think we have a solid data set here ! Let's explore more in the next class.

```

Class excercise: find regions of the clusters that make sense. For example my favorite is that SETDB1 is the methyltransferase that modifies histones to have H3K9me3. Note the two cluster next to each other -- another good sign.

Let's make a heatmap of the promoters -- since it will take a long long time to plot we'll just plot the "super promoters"


```{r}
# We'll start with the promoter peak occurrence and filter down to just those that have high numbers of dbps
high_binders <- peak_occurence_matrix[,colSums(peak_occurence_matrix) > 200]
# Could filter out the rows that have zeros, but we don't have any
min(rowSums(high_binders))

# If you don't have pheatmap, install it.
library(pheatmap)
?pheatmap
# We don't need to be able to read the promoter names and it takes a long time to plot those.
pheatmap(high_binders, show_colnames = FALSE, clustering_distance_rows = "binary", clustering_distance_cols = "binary")

# If we want to just plot the zinc fingers for example, we can reference the num_peaks_df 
num_peaks_df <- read_csv("../../data/num_peaks_df.csv")

zinc_finger_dbd <- num_peaks_df %>% filter(dbd == "C2H2 ZF")

# Index this matrix using the names
zf_matrix <- high_binders[zinc_finger_dbd$dbp,] 
pheatmap(zf_matrix, show_colnames = FALSE, clustering_distance_rows = "binary", clustering_distance_cols = "binary")


# We can also check out the dendrogram for just this matrix.

zf_hclust <- hclust(dist(zf_matrix, method = "binary"))
plot(zf_hclust)
```

