---
title: "15_clustering"
author: "JR"
date: "11/5/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggdendro)
library(GenomicRanges)
library(pheatmap)
source("../../../util/intersect_functions.R")
source("../../../util/plotting_functions.R")
source("../../../util/_setup.R")
```

Today we are going to use hierarchical clustering to identify DBPs with similar binding patterns.

In brief, this involves calculating the distance (euclidian or spearman) between all pairwise 
comparisons of DBP binding profiles. We have made a matrix of 1 and 0 if a DBP is bound at a 
given genome annotation feature (e.g., promoter, gene-body). This is a binary clustering and 
if we cluster a continuous variable (e.g. RNAseq expression levels) then we may want to make 
other considerations for clustering.

We can use those vectors of 1 and 0 to determine how similar the binding profiles are.

Ok let's get started by grabbing the peak_occurrence_matrix from 14_2_our_data_in_ranges.

# loading in promoter peak occurence and convert to matrix
```{r loading in promoter_peak_occurence_df}

#let's look at the promoter peak occurence matrix (takes a min to load)

promoter_peak_occurence_matrix <- read.table("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/12_peak_features/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

# ensuring in matrix format
promoter_peak_occurence_matrix <- as.matrix(promoter_peak_occurence_matrix)
# Ok we see this is a very large matrix of 1 and 0 if a DBP is bound to a promoter.
# We can use these 36,000+ vectors of genes with 483 values of bound (1) or unbound (0)


# let's filter to DBPs with at least 250 binding events or the matrix is a bit sparse.
peak_occurence_matrix <- promoter_peak_occurence_matrix[rowSums(promoter_peak_occurence_matrix) > 250, ]

```

# calculating distnace matrix
Let's get started by finding the 'distance' between each promtoer vector.
We will make a distance matrix of 437 x437 and then use that for clustering.
```{r calculating distance matrix}

# First let's take a look at the matrix

peak_occurence_matrix[1:2,1:40]

# now we can calculate the distance of DBP (we could also do this for promoters)
# let's do a small test first as this can take a while to calculate 451x451!

test <- dist(peak_occurence_matrix[1:2,], method = "euclidean")
test

# There are a number of measures to calculate distance between two vectors and we can check those out in the documentation.
?dist


# let's see how this changes if we use binary as the distance method
# since this is a binary vector we will use a binary distance metric

test2 <- dist(peak_occurence_matrix[1:2,], method = "binary")
test2


# Now let's create distance matrix for each pairwise vector comparison for the whole matrix now.

peak_occurence_dist <- dist(peak_occurence_matrix, method = "binary")

```

# Clustering distance matrix

Now that we have that, we can do hierarchical clustering. In brief, this will cluster 
vectors that have small distances (high correlation), and will iterate that process until
everything is in the same cluster.

```{R hclust of distance matrix}
# let's take a look at hclust

?hclust

# Alright we are ready to cluster everything!

bin_hier <- hclust(peak_occurence_dist, method = "complete")

# let's take a look at what we got and see distance distribution

hist(bin_hier$height)

# we now can plot the information in bin_hier
plot(bin_hier)


# To make this bigger, we can plot it as a pdf the syntax is:
# pdf() tell file path to save and properties then we plot
# finally we call dev.off to take the output into PDF instead of plots.

# We can order this before savin it
# Now let's order the DBPs 
bin_hier$labels[bin_hier$order]

#  !! CHECK WORKING DIRECTORY & structure !!

pdf("figures/dbp_hclust_dendro.pdf", height = 12, width = 70)
plot(bin_hier)
dev.off()

```


# HEATMAPS
Nice so we now have object "bin_hier" that will contain all the cross correlation values across all the samples. 

Now let's use ggdenro package that will plot the branch lengths that indicate how similar two samples are 

```{r ggdendro gram}

# This is a lot of ggplot -- on purpose
# please read through what is happening and parameter usage

 ggdendro::ggdendrogram(bin_hier, rotate = FALSE,  size = 3, 
                       theme_dendro = TRUE) +
   coord_flip() +
   scale_y_continuous() +
   scale_x_continuous(position = "top") +
   scale_x_continuous(breaks = seq_along(bin_hier$labels[bin_hier$order]),
             labels = bin_hier$labels[bin_hier$order], position = "top",
             expand = c(0,0)) +
   theme(axis.text.x = element_text(angle = 90, hjust  = 1)) + 
   theme(axis.text.y = element_text(angle = 0,hjust = 1)) +
   scale_y_reverse(expand = c(0.01, 0)) +
   theme(
     plot.background = element_blank(),
     panel.grid.major = element_blank(),
   panel.grid.minor = element_blank(),
     panel.border = element_blank()
   )

# Nice let's save this using 'ggsave':

ggsave("figures/ggdendro_plot.pdf", height = 50, width = 12, limitsize = F)

```


# Heatmap of lncRNA and mRNA seperately
Now let's compare how lncRNAs and mRNAs cluster differnently?
Strategy: load annotations then index peak_occurrence for specific annotations

#lncRNA promoters
```{r lncRNA promoter clustering}

# We first need to load the .GTF we made for lncRNAs and mRNAs.
# loading lncRNA promoters

lncrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncrna_promoters.gtf")

# loading mRNA promoters
mrna_promoters <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2022/data/genome_features/mrna_promoters.gtf")

# Now we will use indexing to separate peak_occurrence_matrix in lncRNA and mRNA.
# note we are indexing and using indexing in indexing to get what we want.
lncrna_peak_occurence <- peak_occurence_matrix[,lncrna_promoters$gene_id]

# we do the clutering the same as above or we can combine the dist and hclust:
bin_hier_lncrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))

# Now plot with ggdendro
ggdendro::ggdendrogram(bin_hier_lncrna, rotate = TRUE,  size = 3)
 
ggsave("figures/lncrna_hclust_binary_dist.pdf", height = 44, width = 6)
# Cool we can see the Pol II S2 and S5 diff right away again!

```

# mRNA promter heatmap
Now for mRNA

```{R mRNA promoter clustering}

# same strategy used very often:
# indexing into larger file and filtering to a file of stuff you want.
mrna_peak_occurence <- peak_occurence_matrix[,mrna_promoters$gene_id]

# getting the distance matrix for only mRNA promoters  
bin_hier_mrna <- hclust(dist(mrna_peak_occurence, method = "binary"))
 
# plotting with ggdendro
ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)

# saving
ggsave("figures/mrna_hclust_binary_dist.pdf", height = 44, width = 6)

# Cool we are making some good figures -- now let's dig into the biology
# Find an unknown protein cluster that has a known protein gene (hint ZNF)
# read about those genes and see if you could predict what the unknown gene is doing.

```

# Clustering "high binder promoters"
Let's make a heatmap of the promoters -- since it will take a long long time to plot we'll just plot the "super promoters"
```{r clustering high-binding promoters}

# first we filter promoter_peak_occurrence to just those that have high numbers of dbps
# let's, for now, pick 300 DBPs on a given promoter.
# Note another example of indexing a 'master file'  now using a function: colSums
high_binders <- promoter_peak_occurence_matrix[,colSums(promoter_peak_occurence_matrix) > 300]

# Could filter out the rows that have zeros, but we don't have any (min = 1)
min(rowSums(high_binders))

# let's see the diversity of contributions of each DBP to high_binder_promoters
# this means that one DBP is only part of 26 of the 5,661 high binder promoters
hist(rowSums(high_binders))


# Now let's make a heat map with pheatmap, just to switch it up
# also a bit faster -- complex heatmap even faster -- there are many options :)
# library(pheatmap)
?pheatmap


# We don't need to be able to read the promoter names and it takes a long time to plot those.
pheatmap(high_binders, show_colnames = FALSE, clustering_distance_rows = "binary", clustering_distance_cols = "binary")

# Nice let's save that heatmap 

ggsave("figures/high_binder_heatmap.pdf", height = 44, width = 6 )

```

# Clustering ZFN proteins only 
Let's just cluster Zinc Finger proteins (most of unknown function)
There is a theme of take one file and index it with another subset file of interest.

```{R ZFN clustering}

# If we want to just plot the zinc fingers for example, we can reference the num_peaks_df 
# Let's read in num_peaks_df
num_peaks_df <- read_csv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/12_peak_features/num_peaks_df.csv")

# now let's make a zinc_finger_dbd data frame
zinc_finger_dbd <- num_peaks_df %>% filter(dbd == "C2H2 ZF")

# Now we will subset the high binder matrix we made with the
# zinc_finger_dbd object we just made.

zf_matrix <- high_binders[zinc_finger_dbd$dbp,] 

# now the heatmap!

pheatmap(zf_matrix, show_colnames = FALSE, clustering_distance_rows = "binary", clustering_distance_cols = "binary")

ggsave("figures/ZFN_high_binder_heatmap.pdf", height = 44, width = 6 )


# We can also check out the dendrogram for just this matrix.

zf_hclust <- hclust(dist(zf_matrix, method = "binary"))
plot(zf_hclust)

# now let's save it as a PDF

pdf("figures/zfn_clustering.pdf")
zf_hclust <- hclust(dist(zf_matrix, method = "binary"))
plot(zf_hclust)
dev.off()

```

# Adding to num_peaks_df 
While we have all these objects in our environment
let's go ahead and make another data frame.
This will be which DBPs are bound to which promoter.
Thus you can simply look up your favorite promoter 
and retreive all the DBPs bound to it.

```{r DBPs bound to a given promoter}

# We can either read in lncrna_mrna_promoters or merge from here

DBPs_on_promoter <- bind_rows(lncrna_promoters %>% 
                                 as.data.frame(),
                 mrna_promoters %>% 
                    as.data.frame()) %>%
  dplyr::select(gene_id, gene_name)

# This merged the two objects in the environment and made them
# as data frames. How handy is tidyverse!

promoter_dbps <- promoter_peak_occurence_matrix %>%
  as.data.frame() %>%
  rownames_to_column("dbp") %>%
# here we just made a column that has the DBP name
# currently it is a "meta" name
pivot_longer(2:ncol(.), names_to = "gene_id", values_to = "occurrence") %>%
# This now a very long matrix - let's get rid of all zero values
   
  filter(occurrence == 1) %>%
  dplyr::select(-occurrence) %>%
  left_join(promoter_dbps)

# let's write this out it's a very handy DF!
# note saving in 12 where we read in this type of data

write.csv(promoter_dbps, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/12_peak_features/promoter_dbps.csv")

# note this is extremely large but equally useful.
```

# Finding which DBPs are on which promoter
Cool now we can select any gene and see what DBPs are bound?

```{R DBPs on a given promoter}

# In the future if we want to filter on genes we can just read this back in.

promoter_dbps <- read.csv("/Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/13_ChIP_peak_features/promoter_dbps.csv")

# let's see what's on FIRRE

firre_promoter <- promoter_dbps %>%
  filter(gene_name == "FIRRE")

# GAPDH

GAPDH_promoter <- promoter_dbps %>%
  filter(gene_name == "GAPDH")

# trying to find a small binder :)

XIST_promoter <- promoter_dbps %>%
  filter(gene_name == "XIST")

# let's look

XIST_promoter

# Interesting, intuition was correct considering
# this is a male cell line :) but what is bound is interesting
# Note to self STAG1/CTCF is very interesting!
```

Challenge: 

Repeat all code with promoter windows of size +1kb / -.5kb

How do the clustering results change?
