---
title: "16_meta_plots"
author: "JR"
date: "10/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(rtracklayer)
library(tidyverse)
library(ComplexHeatmap)
library(circlize)

source("../../../util/plotting_functions.R")
source("../../../util/intersect_functions.R")
source("../../../util/_setup.R")
```

Today our goal is to make "meta plots" of promoter binding profiles.
In simpler terms: we are going to determine the average binding of:
(A) a given DBP across our promoter windows.

We'll be using the consensus peaks we generated previously (MACS peak calls filtered to those that are overlapping across replicates) and the Gencode gene annotations. This is STEP 0 : loading in required files.

```{r import genome features & file list}

# laoding Gencode genome features
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/gencode.v32.annotation.gtf")


# making a file list of our consensus peaks (filtered at 250 peaks minimum)
fl <- list.files("/Shares/rinn_class/data/CLASS_2022/data/filtered_consensus_peaks/", full.names = TRUE)
consensus_peaks <- lapply(fl, rtracklayer::import)

# adding row names by removing the file path and .bed at end of each file.

names(consensus_peaks) <- gsub("Shares/rinn_class/data/CLASS_2022/data/filtered_consensus_peaks/|.bed|/|_consensus_peaks", "", fl)

# saving this file for later
write_rds(consensus_peaks, "consensus_peaks.rds")

# START POINT: Consensus Peaks
consensus_peaks <- read_rds("consensus_peaks.rds")

```


Overall our goal is to define a "promoter" window of a certain distance
upstream and downstream of a transcription start site (TSS). This will result
in a Xkb (to start 3Kb upstream and downstream) vector representing each base
of the promoter. We will then be able to use this matrix to plot if a peak overlaps
with a given base and do the same for all peaks of a given DBP. 

In the end we will have a verstatil matrix to plot data across promoters. Also we can
use this code to plot other aspects (expression etc) over similar windows. Finally, we can 
use this code to make windows around other features of the geneome (e.g., transposons).

We can accomplish all this in the following 6 steps:


# Step 1: Create promoter windows

Here we're going to create a GRanges object that contains 6kb promoter windows for each gene in the Gencode annotation. First we'll need to filter the Gencode GRanges object to just the genes and then we can use the promoter function from GRanges that will allow us to specify how big of a window we want upstream and downstream of the TSS (you can have asymmetrical windows).

```{r desiging a promoter window of +/- 3Kb from TSS -- probably too big!}

# let's look at where promoters are in gencode_gr
table(gencode_gr$type)

# we want to select the gene category that contains all genes
genes <- gencode_gr[gencode_gr$type == "gene"]

# now we will create promoter windows using genes.
# we call the promoters function in GenomicRanges that defines TSS
# then has a parameter for upstream and downstream sequence:

all_promoters_gr <- promoters(genes, upstream = 3e3, downstream = 3e3)

# results of promoter window definition
table(width(all_promoters_gr))

# now we have a equivalent 6,000 wide vector for each base of a promoter.

```

# Step 2: Transform peaks into a coverage object

In order to calculate what the peak coverage across each promoter is we'll convert the peaks GRanges object which currently holds a range for each peak into a run-length encoded list where 0 represents the genomic coordinates where there is no peak present and 1 represents the locations where a peak is present. The reason for run length encoding is that storing this vector without this compression means that we would be storing a numeric value for each of the 3.2 billion base pairs. This would mean allocating a vector in memory that's ~180 GB -- instead with run-length encoding we're in the ~100 KB range. 

```{r transforming promoter windows into a coverage matrix}

# We will use the coverage function to interpret if there is a 
# given entry of 6,000 wide vectors over each promoter.
# let's first do this for one DBP:

peak_coverage <- coverage(consensus_peaks[["POLR2A"]]) 

# ok now we have a 6,000 length vector with a 1 if a peak
# overlaps the promoter window and 0 if not.

```

### Step 2.1: Some housekeeping to keep our chromosomes straight

This step will accomplish two things: filter out promoters that fall outside the bounds of our coverage vectors and filter out chromosomes that are not in common between the promoters object and the peak coverage object. The reason we need to do this is because the peaks may not extend to the end of each chromosome and therefore there is likely to be promoters that fall outside of the peak coverage vectors -- since we know that there are no peaks on those promoters and therefore they don't give us any extra information about where peaks are relative to promoters we'll filter them out. Also, it creates problems for the Views command that we'll use to subset the coverage vectors to just the promoter windows.

```{r peak_coverage of all promoter windows : from GRanges to Iranges}

# This is the length of each run-length encoded vector in the peak_coverage object
# If the last peak on each chromosome falls near the end of that chromosome then
# these lengths will be approximately the length of the chromosomes.
coverage_length <- elementNROWS(peak_coverage)

# This will create a GRanges object where there is one range per chromosome
# and it is the width of the coverage vector -- we can use these ranges to 
# filter the promoters falling outside of these boundaries in the next step.
# Each DBP will be different here.

coverage_gr <- GRanges(seqnames = names(coverage_length),
                       IRanges(start = rep(1, length(coverage_length)), 
                               end = coverage_length))

# Okay, now we're all ready to filter out those promoters that fall beyond the bounds of the 
# coverage vector. 
all_promoters_gr <- subsetByOverlaps(all_promoters_gr, 
                                  coverage_gr, 
                                  type="within", 
                                  ignore.strand=TRUE)


# This creates a vector of the chromosomes that are in common between the peak_coverage and promoters.
# Keeping track of this will fix a similar problem to the step above since this DBP may not
# bind on all chromosomes. 
chromosomes <- intersect(names(peak_coverage), unique(as.character(seqnames(all_promoters_gr))))


# We can also ensure they're in the same order and contain the same chromosomes
# by indexing with this vector
peak_coverage <- peak_coverage[chromosomes]


# In order to match the list format of the peak_coverage object
# we'll also coerce the GRanges object into an IntegerRangesList.
# If you recall, one of the main features of GRanges object is capturing
# the chromosome information -- when converting to an IRanges list, 
# each chromosome will be represented by a named element in the list.

all_promoters_ir <- as(all_promoters_gr, "IntegerRangesList")[chromosomes]

# nice now we have all the promoter windows as an Iranges object
```

## Step 3: Subset the peak coverage vector to just the promoter windows

Here we'll use the Views function to mask the peak coverage object everywhere but in the windows of the promoters.

```{r reducing data to jsut Iranges of promters and peak coverage }

# let's look at vies
?Views
# ok so Views can store and access all the promoter vectors
# We will end up with an index of all 'promoter windows'

# indexing all the promter window vectors (6Kb)
promoter_peak_view <- Views(peak_coverage, all_promoters_ir)

#let's look at what we got
promoter_peak_view
# Note that these are still in run-length encoding format.

```

## Step 4: Contstruct a matrix of the coverage values of each promoter region

We'll not just convert the run-length encoding vectors to actual vectors -- note how much larger the object becomes when represented as vectors (use object.size function). Then we'll row bind the vectors into one matrix.

```{r converting RLE to large matrix extracting views to 1 and 0}

# we will use viewApply to extract if there was a 1 or 0 in the RLE vector for an overlap.
# This essentially will convert the RLE to 6000 length vecotrs of promoter windows -- with
# a 1 for binding event and 0 if not.

promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))

# We are now going to convert promoter_peak_view to a matrix
# we will do this by rowbinding the Irange list 

# Base R
promoter_peak_matrix <- do.call("rbind", promoter_peak_view)

# TODO this doesn't work but the above does
# test this 
promoter_peak_matrix <- promoter_peak_view %>%
  bind_rows()

# now we have a matrix of nrow_promoters by 6,000 cols
```

## Step 5: Align the positive and negative strand promoters

Since the genes that are transcribed from the minus strand will have their upstream and downstream values flipped relative to the plus strand promoters, we need to reverse those vectors so that upstream and downstream values are consistent.

```{r setting all promoters to the same direction}

# We're just going to flip one strand because we want to get them in the same orientation

#TODO understand this?

minus_idx <- which(as.character(strand(all_promoters_gr)) == "-")
promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx, ncol(promoter_peak_matrix):1]

# Now we can get rid of the rows that have no peaks.
promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 0,]

# now we have a matrix of only promoter windows with at least 1 base overlap with
# a promoter region for a given DBP. 

```

## Step 6: Sum the columns, normalize, and plot

To summarize this matrix, we'll sum up the number of binding events at each position in this 6kb window. This vector represents the overall peak coverage of each posistion, for the purpose of visualizing this, we'll normalize by the total coverage so that the area under the curve in the plot sums to one. 

```{r peak coverage for a given DBP over all promoter windows}

# First, calculating the number of peak overlaps at each base of promoter window.

# columns = 6,000 each a base in promoter window.
# rows = all promoter windows.
# sum of cols (colSums) = how many peak overlaps there were at each base (cols).

peak_sums <- colSums(promoter_peak_matrix)

# Normalization of binding events at each base of promoter window.
# we will divide the number of peak sums at each base by the total
# number of peak binding events.

# noramkization of binding events to 1 across promoter window
peak_dens <- peak_sums/sum(peak_sums)

# Create a data frame in order to plot this. 
metaplot_df <- data.frame(x = -3e3:(3e3-1), dens = peak_dens)

# now we have a data frame of cols as base in promoter window
# rows as each promoter window. Normalize by percentage of total binding events.
```


Plotting POLR2A binding coverage across all bound promoters
```{R POLR2A binding coverage across all bound promoters}

# plotting the data in the metaplot_df
ggplot(metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)

# Nice, POL2RA has normal distribution of binding events
# around the center of the promtoer window (TSS).

# We have now done this for one DBP and can write the above as a function
# to run across all DBPs later. More soon on that front!
```

This plot is pretty basic, so let's add some more meangingful legends and a title.
```{r Better POLR2A meta plot}

# Plot it with ggplot geom_line
ggplot(metaplot_df, aes(x = x, y = dens)) + 
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  ggtitle("POLR2A Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-3000, 0, 3000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency") +
ggsave("figures/POLR2A_promoter_metaplot.pdf")
```

# Let's put all this into a funciton so we can call it later conveniently.

```{r}
profile_tss <- function(peaks, 
                        promoters_gr,
                        upstream = 3e3,
                        downstream = 3e3) {
  

  peak_coverage <- coverage(peaks)
  
  coverage_length <- elementNROWS(peak_coverage)
  coverage_gr <- GRanges(seqnames = names(coverage_length),
                         IRanges(start = rep(1, length(coverage_length)), 
                                 end = coverage_length))
  
  promoters_gr <- subsetByOverlaps(promoters_gr, 
                                       coverage_gr, 
                                       type="within", 
                                       ignore.strand=TRUE)
  chromosomes <- intersect(names(peak_coverage), 
                           unique(as.character(seqnames(promoters_gr))))
  peak_coverage <- peak_coverage[chromosomes]
  
  promoters_ir <- as(promoters_gr, "IntegerRangesList")[chromosomes]
  
  promoter_peak_view <- Views(peak_coverage, promoters_ir)
  
  promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))
  promoter_peak_matrix <- do.call("rbind", promoter_peak_view)
  
  minus_idx <- which(as.character(strand(promoters_gr)) == "-")
  promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx,
                                                           ncol(promoter_peak_matrix):1]
  
  promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 1,]
  
  peak_sums <- colSums(promoter_peak_matrix)
  peak_dens <- peak_sums/sum(peak_sums)
  
  metaplot_df <- data.frame(x = -upstream:(downstream-1),
                            dens = peak_dens)
  
  return(metaplot_df)
}
```

## Use this function to make separate plots for lncRNA and mRNA

First we'll create separate objects for lncRNA promoters and mRNA promoters, then we'll supply each of these to the new function we just made.

```{r}
# lncRNA promoter profiles
lncrna_genes <- genes[genes$gene_type == "lncRNA"]
lncrna_promoters <- promoters(lncrna_genes, upstream = 3e3, downstream = 3e3)


# TODO keep this and add below to make more plots
lncrna_metaplot_profile <- profile_tss(consensus_peaks[["POLR2A"]], lncrna_promoters)

# mRNA promoter profiles
mrna_genes <- genes[genes$gene_type == "protein_coding"]
mrna_promoters <- promoters(mrna_genes, upstream = 3e3, downstream = 3e3)

mrna_metaplot_profile <- profile_tss(consensus_peaks[["POLR2A"]], mrna_promoters)

# We can row bind these dataframes so that we can plot them on the same plot
mrna_metaplot_profile$gene_type <- "mRNA"
lncrna_metaplot_profile$gene_type <- "lncRNA"
combined_metaplot_profile <- bind_rows(mrna_metaplot_profile, lncrna_metaplot_profile)

ggplot(combined_metaplot_profile, 
       aes(x = x, y = dens, color = gene_type)) +
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  ggtitle("POLR2A Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-3000, 0, 3000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency") + 
  scale_color_manual(values = c("#424242","#a8404c"))
```

```{r creating coverage matrix for all DBPs }

# Let's now run this for all off the DBPs and compile it into one data frame.
# Let's first define an empty data.frame to which we can row_bind each new one created.
metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# Writting a for loop to calculate promter coverage for all DBPs:

for(i in c(1:390, 392:length(consensus_peaks))) {
  # we skipped 391 as it breaks (only 19 peaks ?)
  print(names(consensus_peaks)[[i]])
  tmp_df <- profile_tss(consensus_peaks[[i]], promoters_gr = all_promoters_gr)
  tmp_df$dbp <- names(consensus_peaks)[[i]]
  metaplot_df <- bind_rows(metaplot_df, tmp_df)
  
}

# Let's save the df to make all meta plots of all DBPs !
write_csv(metaplot_df, "metaplot_df.csv")

write_rds(metaplot_df, "metaplot_df.rds")

metaplot_df <- read_rds("metaplot_df.rds")

#TODO plot some meta plots here TP53, YY1




```

```{r filtering consensus peaks to those > 250 peaks per DBP}


# Let's filter out those peaks that don't pass the threshold, since
# in particular here it will make a difference, since the 
# average profile will be very lumpy for those DBPs with few peaks.
passing_peaks <- names(consensus_peaks)[sapply(consensus_peaks, length) > 250]

# Let's cluster the metaplots df. First we need to turn it into a matrix.
metaplot_df <- read_rds("/Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/16_meta_plots/metaplot_df.rds") %>%
  filter(dbp %in% passing_peaks) %>%
  distinct()


# Pivot wider into a matrix
# careful you do have duplicated rows
# we shouldn't since we ran distinct above.

metaplot_matrix <- metaplot_df %>% 
  pivot_wider(names_from = x, values_from = dens) %>%
  column_to_rownames("dbp") %>%
  as.matrix()
dim(metaplot_matrix)


# Z-Scale the rows
mm_scaled <- t(scale(t(metaplot_matrix)))


metaplot_hclust <- hclust(dist(mm_scaled), method = "complete")

# Plot the dendrogram
pdf("figures/tss_profile_dendrogram.pdf", height = 10, width = 27)
par(cex=0.3)
plot(metaplot_hclust)
dev.off()

# Cut the tree to make some clusters
clusters <- cutree(metaplot_hclust, h = 0.01)

# Let's make a heatmap now.
col_fun <- colorRamp2(c(-3, 0, 3), c("#5980B3", "#ffffff", "#B9605D"))
split <- data.frame(split = c(rep("-3kb",3000), rep("+3kb", 3000)))

pdf("figures/tss_profile_heatmap.pdf", height = 35, width = 10)
Heatmap(mm_scaled, cluster_columns = FALSE, col = col_fun, border = TRUE, 
        show_column_names = FALSE,
        use_raster = TRUE,
        column_split = split,
        column_gap = unit(0, "mm"),row_names_gp = gpar(fontsize = 7))
dev.off()

par(cex = 1)
# Let's plot some individual TSS profiles
plot_tss_profile(metaplot_matrix, "H3K4me1", save_pdf = TRUE)
plot_tss_profile(metaplot_matrix, "TCF12", save_pdf = TRUE)

# Look at what the average profiles look like in each cluster.

```

