---
title: "16_meta_plots"
author: "JR"
date: "11/26/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(rtracklayer)
library(tidyverse)
library(ComplexHeatmap)
library(circlize)

source("../../../util/plotting_functions.R")
source("../../../util/intersect_functions.R")
source("../../../util/_setup.R")
```

Today our goal is to make "meta plots" of promoter binding profiles.
In simpler terms: we are going to determine the average binding of:
(A) a given DBP across our promoter windows.

We'll be using the consensus peaks we generated previously (MACS peak calls filtered to those that are overlapping across replicates) and the Gencode gene annotations. This is STEP 0 : loading in required files.

# Loading in data needed for meta plots
```{r import genome features & file list}

# Loading in filtered consensus peaks
fl <- list.files("/scratch/Shares/rinnclass/CLASS_2022/data/filtered_consensus_peaks", 
                                             pattern = "*.bed",
                                             full.names = TRUE)

# lappy to import each file in file list 
filtered_consensus_peaks <- lapply(fl, rtracklayer::import)


names(filtered_consensus_peaks) <- sapply(filtered_consensus_peaks, function(x){
  unlist(strsplit(x$name, "_"))[[1]]
})

```

Overall our goal is to define a "promoter" window of a certain distance
upstream and downstream of a transcription start site (TSS). This will result
in a Xkb (to start 3Kb upstream and downstream) vector representing each base
of the promoter. We will then be able to use this matrix to plot if a peak overlaps
with a given base and do the same for all peaks of a given DBP. 

In the end we will have a versatile matrix to plot data across promoters. Also we can
use this code to plot other aspects (expression etc) over similar windows. Finally, we can 
use this code to make windows around other features of the geneome (e.g., transposons).

We can accomplish all this in the following 6 steps:


# Step 1: Create promoter windows

Here we're going to create a GRanges object that contains 6kb promoter windows for each gene in the Gencode annotation. First we'll need to filter the Gencode GRanges object to just the genes and then we can use the promoter function from GRanges that will allow us to specify how big of a window we want upstream and downstream of the TSS (you can have asymmetrical windows).

```{r desiging a promoter window of +/- 3Kb from TSS -- probably too big!}

# First thing to do is load in all the promoters we made in 12_ 
# Loading in the promoter regions for lncRNAs and mRNAs 
all_promoters_gr <- rtracklayer::import("/path/gene_annotation_files/lncRNA_mrna_promoters.gtf")

# all_promoters_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/11_consensus_peaks/gene_annotation_files/lncRNA_mrna_promoters.gtf")

```

# Step 2: Transform chip-peaks into a coverage object

In order to calculate what the peak coverage across each promoter is we'll convert the peaks GRanges object which currently holds a range for each peak into a run-length encoded list where 0 represents the genomic coordinates where there is no peak present and 1 represents the locations where a peak is present. The reason for run length encoding is that storing this vector without this compression means that we would be storing a numeric value for each of the 3.2 billion base pairs. This would mean allocating a vector in memory that's ~180 GB -- instead with run-length encoding we're in the ~100 KB range. 

```{r transforming promoter windows into a coverage matrix}

# We will use the coverage function to interpret if there is an overlap 
# at given entry of 6,000 entries for each base of the 'promoter window

?coverage

# random but there needs to be spaces before and after gene_id ??
peak_coverage <- coverage(filtered_consensus_peaks[["POLR2A"]]) 
# ok now we have a 6,000 length vector with 1 for overlap or 0
View(peak_coverage)
# note each chromosome is a seperate list of ranges of overlaps or non-overlaps

```

### Step 2.1: Some housekeeping to keep our chromosomes straight

This step will accomplish two things: filter out promoters that fall outside the bounds of our coverage vectors and filter out chromosomes that are not in common between the promoters object and the peak coverage object. The reason we need to do this is because the peaks may not extend to the end of each chromosome and therefore there is likely to be promoters that fall outside of the peak coverage vectors -- since we know that there are no peaks on those promoters and therefore they don't give us any extra information about where peaks are relative to promoters we'll filter them out. Also, it creates problems for the Views command that we'll use to subset the coverage vectors to just the promoter windows.

```{r peak_coverage of all promoter windows : from GRanges to Iranges}

# This is the length of each run-length encoded vector in the peak_coverage object
?elementNROWS
# elementNROWs can calculate coverage in RLE
coverage_length <- elementNROWS(peak_coverage)

# This will create a Ranges object where there is one range per chromosome
# and it is the width of the coverage vector that covers all promoters-- 

# here are the components:
names(coverage_length)
# chromosome names

length(coverage_length)
# 24 chromosomes

# we can use the coverage length ranges to 
# filter the promoters falling outside of these boundaries:
# basically turning coverage_lenth to GRanges
coverage_gr <- GRanges(seqnames = names(coverage_length),
                       IRanges(start = rep(1, length(coverage_length)), 
                               end = coverage_length))

# Okay, now we're all ready to filter out those promoters that fall beyond the bounds of the 
# coverage vector. 

# to do this we will use:
?subsetByOverlaps

all_promoters_gr <- subsetByOverlaps(all_promoters_gr, 
                                  coverage_gr, 
                                  type="within", 
                                  ignore.strand=TRUE)



# IMPORTANT STEP: this will take names of chr with peaks and intersect with all chr
# Then after this we will run coverage on our chromosome object. 
# If we didn't do this an error would occur if a chromosome wasn't represented

# this is what we are intersecting:
names(peak_coverage)
# with
seqnames(all_promoters_gr)
# note we would have lossed the mitochondrial chromosome!
chromosomes <- intersect(names(peak_coverage), unique(as.character(seqnames(all_promoters_gr))))


# We can also ensure they're in the same order and contain the same chromosomes
# by indexing with this vector
peak_coverage <- peak_coverage[chromosomes]


# In order to match the list format of the peak_coverage object
# we'll also coerce the GRanges object into an IntegerRangesList.
# If you recall, one of the main features of GRanges object is capturing
# the chromosome information -- when converting to an IRanges list, 
# each chromosome will be represented by a named element in the list.
?IntegerRangesList
# using base R here with 'as' a coericing function. 
all_promoters_ir <- as(all_promoters_gr, "IntegerRangesList")[chromosomes]

# nice now we have all the promoter windows as an Iranges object
```

## Step 3: Subset the peak coverage vector to just the promoter windows

Here we'll use the Views function to mask the peak coverage object everywhere but in the windows of the promoters.

```{r reducing data to jsut Iranges of promters and peak coverage }

# let's look at views
?Views
# ok so Views can store and access all the promoter vectors
# We will end up with an index of all 'promoter windows'

# indexing all the promter window vectors (6Kb)
promoter_peak_view <- Views(peak_coverage, all_promoters_ir)

#let's look at what we got
promoter_peak_view
# Note that these are still in run-length encoding format.

```

## Step 4: Contstruct a matrix of the coverage values of each promoter region

We'll not just convert the run-length encoding vectors to actual vectors -- note how much larger the object becomes when represented as vectors (use object.size function). Then we'll row bind the vectors into one matrix.

```{r converting RLE to large matrix extracting views to 1 and 0}

# we will use viewApply to extract if there was a 1 or 0 in the RLE vector for an overlap.
# This essentially will convert the RLE to 6000 length vecotrs of promoter windows -- with
# a 1 for binding event and 0 if not.

?viewApply
# This performs functions on each of the vectors in view list.
# kind of like mutate in vector space (note the matrix are ncol and 6,000 rows)

promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))

# Let's look (note rows turned into cols)
promoter_peak_view

# We are now going to convert promoter_peak_view to a matrix
# we will do this by rowbinding the Irange list that viewApply
# turned into a vector.

# Base R do.call to call "rbind" function
promoter_peak_matrix <- do.call("rbind", promoter_peak_view)
# good to note in a matrix sometimes to use do.call and "rbind"

dim(promoter_peak_matrix)
# now we have a matrix of nrow_promoters by 6,000 cols

```

## Step 5: Align the positive and negative strand promoters

Since the genes that are transcribed from the minus strand will have their upstream and downstream values flipped relative to the plus strand promoters, we need to reverse those vectors so that upstream and downstream values are consistent.

```{r setting all promoters to the same direction}

# We're just going to flip one strand because we want to get them in the same orientation
# It really wouldn't matter for a sysmetric window size but good to keep in mind the strand
# could matter.

minus_idx <- which(as.character(strand(all_promoters_gr)) == "-")
# taking only minus strand rows

# we want to flip so instead of 1:ncol we will do ncol:1 (currenlty 6,000 - 1)
# we want position 1 - 6000

promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx, ncol(promoter_peak_matrix):1]

# Now we can get rid of the rows that have no peaks.
promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 0,]

# now we have a matrix of only promoter windows with at least 1 base overlap with
# a promoter region for a given DBP. 

```

## Step 6: Sum the columns, normalize, and plot

To summarize this matrix, we'll sum up the number of binding events at each position in this 6kb window. This vector represents the overall peak coverage of each posistion, for the purpose of visualizing this, we'll normalize by the total coverage so that the area under the curve in the plot sums to one. 

```{r peak coverage for a given DBP over all promoter windows}

# First, calculating the number of peak overlaps at each base of promoter window.

# columns = 6,000 each a base in promoter window.
# rows = all promoter windows.
# sum of cols (colSums) = how many peak overlaps there were at each base (cols).

peak_sums <- colSums(promoter_peak_matrix)

# Normalization of binding events at each base of promoter window.
# we will divide the number of peak sums at each base by the total
# number of peak binding events.

# noramkization of binding events to 1 across promoter window
peak_dens <- peak_sums/sum(peak_sums)

# Create a data frame in order to plot this. 
metaplot_df <- data.frame(x = -3e3:(3e3-1), dens = peak_dens)

# now we have a data frame of cols as base in promoter window
# rows as each promoter window. Normalize by percentage of total binding events.
```

# Plot POLR21 metaplot
Plotting POLR2A binding coverage across all bound promoters
```{R POLR2A binding coverage across all bound promoters}

# plotting the data in the metaplot_df
ggplot(metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)

# Nice, POL2RA has normal distribution of binding events
# around the center of the promtoer window (TSS).

# We have now done this for one DBP and can write the above as a function
# to run across all DBPs later. More soon on that front!
```

# fancy POLR2A metaplot
This plot is pretty basic, so let's add some more meangingful legends and a title.
```{r Better POLR2A meta plot}

# Plot it with ggplot geom_line
ggplot(metaplot_df, aes(x = x, y = dens)) + 
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  ggtitle("POLR2A Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-3000, 0, 3000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency")
ggsave("figures/POLR2A_promoter_metaplot.pdf")
```

# creating a meta plot function for all DBPs
Let's put all this into a funciton so we can call it later conveniently.

```{r establishing the profile_tss function}


# setting the parameters
profile_tss <- function(peaks, 
                        promoters_gr,
                        upstream = 3e3,
                        downstream = 3e3) {
  
# performing coverage function 
  peak_coverage <- coverage(peaks)

# keeping track of overlaps in RLE
  coverage_length <- elementNROWS(peak_coverage)

# Defining a GRanges of the promter window
  coverage_gr <- GRanges(seqnames = names(coverage_length),
                         IRanges(start = rep(1, length(coverage_length)), 
                                 end = coverage_length))
  
# defining the promoters 
  promoters_gr <- subsetByOverlaps(promoters_gr, 
                                       coverage_gr, 
                                       type="within", 
                                       ignore.strand=TRUE)

# making sure the chromosomes represented are used (prevent error if chr is missing)
  chromosomes <- intersect(names(peak_coverage), 
                           unique(as.character(seqnames(promoters_gr))))

# arranging chromosomes in the same order
  peak_coverage <- peak_coverage[chromosomes]

# converting to InterRangesList
  promoters_ir <- as(promoters_gr, "IntegerRangesList")[chromosomes]

# creating a views object for promoter coverage (because in RLE)
  promoter_peak_view <- Views(peak_coverage, promoters_ir)

# turning into a vector with ViewApply (because in RLE keeping track of where overlaps are)
  promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))

# binding each of the view vectors
  promoter_peak_matrix <- do.call("rbind", promoter_peak_view)

# grabing and reversing promoters on the - strand
  minus_idx <- which(as.character(strand(promoters_gr)) == "-")
  
# reversing the order from 6,000 - 1 to 1- 6000
  promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx,
                                                           ncol(promoter_peak_matrix):1]

# eliminating promoters with no binding 
  promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 1,]

# summing all the vectors of a given DBP to the promoter window
  peak_sums <- colSums(promoter_peak_matrix)

# calculating the density at each position in the promoter window
  peak_dens <- peak_sums/sum(peak_sums)

# making it go from -3K to + 3K and creating a df
  metaplot_df <- data.frame(x = -upstream:(downstream-1),
                            dens = peak_dens)
  
  return(metaplot_df)
}
```

# Use this function to make separate plots for lncRNA and mRNA

First we'll create separate objects for lncRNA promoters and mRNA promoters, then we'll supply each of these to the new function we just made.

```{r making seperate metaplots for lncRNAs and mRNAs}


# First we will need some gene annotation data from gencode_gr :) 
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/gencode.v32.annotation.gtf")

# Indexing gencode_gr to grab "genes"
genes <- gencode_gr[gencode_gr$type == "gene"]

# lncRNA promoter profiles

lncrna_genes <- genes[genes$gene_type == "lncRNA"]

lncrna_promoters <- promoters(lncrna_genes, upstream = 3e3, downstream = 3e3)


# Now we can use profile_tss
# POLR2A meta-plot / feel free to pick any other gene
lncrna_metaplot_profile <- profile_tss(filtered_consensus_peaks[["POLR2A"]], lncrna_promoters)

# test plot 
ggplot(lncrna_metaplot_profile, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)

# mRNA promoter profiles
mrna_genes <- genes[genes$gene_type == "protein_coding"]

mrna_promoters <- promoters(mrna_genes, upstream = 3e3, downstream = 3e3)

mrna_metaplot_profile <- profile_tss(filtered_consensus_peaks[["POLR2A"]], mrna_promoters)

ggplot(mrna_metaplot_profile, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)

# so polR2a looks similar on lncRNA and mRNA promoters

# We can row bind these dataframes so that we can plot them on the same plot
mrna_metaplot_profile$gene_type <- "mRNA"
lncrna_metaplot_profile$gene_type <- "lncRNA"
combined_metaplot_profile <- bind_rows(mrna_metaplot_profile, lncrna_metaplot_profile)

ggplot(combined_metaplot_profile, 
       aes(x = x, y = dens, color = gene_type)) +
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  ggtitle("POLR2A Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-3000, 0, 3000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency") + 
  scale_color_manual(values = c("#424242","#a8404c"))

# combinig on the same plot shows the difference in binding !
```

Making coverage matrix for all DBPs across mRNA and lncRNA promoters
# NOTE THIS TAKES ~ 3hrs to run !

```{r creating coverage matrix for all DBPs }

# Let's now run this for all off the DBPs and compile it into one data frame.
# Let's first define an empty data.frame to which we can row_bind each new one created.
metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# Writting a for loop to calculate promoter coverage for all DBPs:

for(i in c(1:390, 392:length(filtered_consensus_peaks))) {
  # we skipped 391 as it breaks (only 19 peaks ?)
  print(names(filtered_consensus_peaks)[[i]])
  tmp_df <- profile_tss(filtered_consensus_peaks[[i]], promoters_gr = all_promoters_gr)
  tmp_df$dbp <- names(filtered_consensus_peaks)[[i]]
  metaplot_df <- bind_rows(metaplot_df, tmp_df)
  
}

# write_rds(metaplot_df, "metaplot_df.rds")

write_rds(metaplot_df, "metaplot_df_test.rds")

# some more filtering and writing out:
metaplot_df <- read_rds("metaplot_df.rds")

# filtering to be sure 
passing_peaks <- names(filtered_consensus_peaks)[sapply(filtered_consensus_peaks, length) > 250]

metaplot_filtered_df <- metaplot_df %>%
  filter(dbp %in% passing_peaks) 
    

# writing out
# MAKE SURE IN RIGHT WORKING DIRECTORY
write_rds(metaplot_filtered_df, "metaplot_filtered_df.rds")

setdiff(unique(metaplot_df$dbp), unique(metaplot_filtered_df$dbp))

metaplot_filtered_df <- read_rds("metaplot_filtered_df.rds")


```

##### START POINT #####

```{R start point for creating all meta plots}

```

##### START POINT ########
```{r START POINT TO CREATE META_PLOTS}

# read in metaplot_df
metaplot_df <- read_rds("metaplot_df.rds")

# Pivot wider into a matrix
# careful we don't have duplicated rows
# we shouldn't since we ran distinct above.

metaplot_filtered_matrix <- metaplot_df %>% 
  pivot_wider(names_from = x, values_from = dens) %>%
  column_to_rownames("dbp") %>%
  as.matrix()

# Get rid of NA rows. Not enough data to calculate a metaplot.
nrow(metaplot_filtered_matrix)
dim(metaplot_filtered_matrix)

# Let's see how many have NA values
table(rowSums(is.na(metaplot_filtered_matrix)))

# Let's filter out those with NAs
metaplot_filtered_matrix <- metaplot_filtered_matrix[rowSums(is.na(metaplot_filtered_matrix)) == 0,]


#TODO make diagram of rows and cols
# Z-Scale the rows for clustering (not needed for profile plots)
mm_scaled <- t(scale(t(metaplot_filtered_matrix)))

# clustering
metaplot_hclust <- hclust(dist(mm_scaled), method = "complete")

# plotting
plot(metaplot_hclust)

# Plot the dendrogram
pdf("figures/tss_profile_dendrogram.pdf", height = 10, width = 27)
par(cex=0.3)
plot(metaplot_hclust)
dev.off()

# Cut the tree to make some clusters let's start with 10
clusters <- cutree(metaplot_hclust, k=10)
table(clusters)

# Let's shrink to 4 clusters based on table
clusters <- cutree(metaplot_hclust, k = 3)
table(clusters)


# Let's make a heatmap now.
col_fun <- colorRamp2(c(-3, 0, 3), c("#5980B3", "#ffffff", "#B9605D"))
split <- data.frame(split = c(rep("-3kb",3000), rep("+3kb", 3000)))

# pdf("figures/tss_profile_heatmap.pdf", height = 35, width = 10)
Heatmap(mm_scaled, cluster_columns = FALSE, col = col_fun, border = TRUE, 
        show_column_names = FALSE,
        use_raster = TRUE,
        column_split = split,
        column_gap = unit(0, "mm"),row_names_gp = gpar(fontsize = 7))
# dev.off()

# Now a heatmap with only 4 clusters since most of the data is the same.

Heatmap(mm_scaled[clusters == 3,], cluster_columns = FALSE, col = col_fun, border = TRUE, 
        show_column_names = FALSE,
        use_raster = TRUE,
        column_split = split,
        column_gap = unit(0, "mm"),row_names_gp = gpar(fontsize = 7))
par(cex = 1)


```

# Looking at some example meta-plots
Here is some fun code to pick your favorite DBP and check out the profile.
I chose those DBPs that seemed to have the most different binding profiles.

```{r meta-plots of most different binding profiles by clustering}

# H3K4me1 H4K20me1 H3K36me3 MBD4:

# H3K4me1: shows inverse pattern to most with depletion at TSS

plot_tss_profile(metaplot_filtered_matrix, "H3K4me1", save_pdf = TRUE)

# H3K36me3: also depleted at TSS as expected marks genebodies 

plot_tss_profile(metaplot_filtered_matrix, "H3K36me3", save_pdf = TRUE)


# KAT2B, H3K9me3, NFIC cluster:
# H3K9me3: seems to go into gene body
plot_tss_profile(metaplot_filtered_matrix, "H3K9me3", save_pdf = TRUE)


# EP300: shows a very wide profile.
plot_tss_profile(metaplot_matrix, "EP300", save_pdf = TRUE)

```



















```{r start point reading in metaplot_filtered_df}

#  why doesn't it work reading in the .rds
# This is reading in the filtered even though we are 
# setting to meta_plot_df

# this somehow doesn't work when read in -- but does work
# TODO works when in environment -- think I will try and save matrix and see if it works.

# metaplot_filtered_matrix_test <- read_rds("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/16_meta_plots/metaplot_filtered_matrix.rds")

# hmm <- read_tsv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/16_meta_plots/metaplot_filtered.tsv")

#metaplot_df <- read_rds("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/16_meta_plots/metaplot_filtered_df.rds") 

# metaplot_df <- subset(metaplot_df, select = -c(x))

```


# clustering dbp by binding profile

```{R clustering each DBP by their binding profile over promoters}

# Let's cluster the metaplots df. First we need to turn it into a matrix.


# Pivot wider into a matrix
# careful we don't have duplicated rows
# we shouldn't since we ran distinct above.

# metaplot_filtered_matrix <- metaplot_filtered_df %>% 
#   pivot_wider(names_from = x, values_from = dens) %>%
#   column_to_rownames("dbp") %>%
#   as.matrix()
# dim(metaplot_filtered_matrix)

# ?hclust
# # Z-Scale the rows for clustering (not needed for profile plots)
# mm_scaled <- t(scale(t(metaplot_filtered_matrix)))
# 
# # clustering
# metaplot_hclust <- hclust(dist(mm_scaled), method = "complete")
# 
# # plotting
# plot(metaplot_hclust)

# Plot the dendrogram
# pdf("figures/tss_profile_dendrogram.pdf", height = 10, width = 27)
# par(cex=0.3)
# plot(metaplot_hclust)
# dev.off()

# Cut the tree to make some clusters
# clusters <- cutree(metaplot_hclust, k = 3)
# summary(metaplot_hclust$height)
# # Let's make a heatmap now.
# col_fun <- colorRamp2(c(-3, 0, 3), c("#5980B3", "#ffffff", "#B9605D"))
# split <- data.frame(split = c(rep("-3kb",3000), rep("+3kb", 3000)))
# 
# pdf("figures/tss_profile_heatmap.pdf", height = 35, width = 10)
# Heatmap(mm_scaled, cluster_columns = FALSE, col = col_fun, border = TRUE, 
#         show_column_names = FALSE,
#         use_raster = TRUE,
#         column_split = split,
#         column_gap = unit(0, "mm"),row_names_gp = gpar(fontsize = 7))
# dev.off()
# 
# 
# Heatmap(mm_scaled[clusters == 3,], cluster_columns = FALSE, col = col_fun, border = TRUE, 
#         show_column_names = FALSE,
#         use_raster = TRUE,
#         column_split = split,
#         column_gap = unit(0, "mm"),row_names_gp = gpar(fontsize = 7))
# table(clusters)
# 
# par(cex = 1)

```

```{r}
# table

# write.table(metaplot_filtered_matrix, "csv_metaplot_filtered_natrix.tsv" )
# test.tsv <- read.table("csv_metaplot_filtered_natrix.tsv")
# 
# write.table(metaplot_filtered_matrix, "metaplot_filtered.rds" )
# 
# write.table(metaplot_filtered_matrix, "metaplot_filtered.tsv" )
# 
# test_read_table <- read.table("metaplot_filtered.tsv")
# 
# test_read_table <- read.table("metaplot_filtered.rds")

```

# convert to metaplot matrix
```{r converting metaplot_df to matrix}
# 
# # Pivot wider into a matrix
# # careful we don't have duplicated rows
# # we shouldn't since we ran distinct above.
# 
# metaplot_filtered_matrix <- metaplot_df %>% 
#   pivot_wider(names_from = x, values_from = dens) %>%
#   column_to_rownames("dbp") %>%
#   as.matrix()
# 
# # Get rid of NA rows. Not enough data to calculate a metaplot.
# nrow(metaplot_filtered_matrix)
# metaplot_filtered_matrix <- metaplot_filtered_matrix[rowSums(is.na(metaplot_filtered_matrix)) == 0,]
# 
# # metaplot_filtered_matrix <- metaplot_filtered_df %>% 
# #   pivot_wider(names_from = x, values_from = dens) %>%
# #   column_to_rownames("dbp") %>%
# #   as.numeric() %>%
# #   as.matrix() 






```
