---
title: "18_Rnaseq_part_II"
author: "JR"
date: "11/12/2021"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(tximport)
library(DESeq2)
library(ggrepel)
library(pheatmap)
```

Let's do some RNAseq !

Goal: Align reads and get the number of reads in each gene in the genome using nf-core/rnaseq 

To use this new version of the nf-core/rnaseq pipeline (v3.4) 
So what we need is

design.csv
nextflow.config file
run.sh


Let's start with the config file -- very similar to nf-core/chipseq
```{bash}
process {
  executor='slurm'
  queue='short'
  memory='16 GB'
  maxForks=10
}
```


Now the run.sh file that gives nf-core/rnaseq all the instructions for nextflow to task.
```{bash run.sh file}

# Here are all the slurm instructions:


#!/bin/bash
#SBATCH -p long
#SBATCH --job-name=hepg2-fraction-RNAseq
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=x@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=6gb
#SBATCH --time=100:00:00
#SBATCH --output=nextflow.out
#SBATCH --error=nextflow.err

pwd; hostname; date
echo "You've requested $SLURM_CPUS_ON_NODE core."

module load singularity/3.1.1

nextflow run nf-core/rnaseq -r 3.4 \
-resume \
-profile singularity \

# Now the inputs for nextflow.

# Firs the design file that we just made.
--input /Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/17_API_to_RNAseq/design.csv \

# The genome file
--fasta /scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/GRCh38.p13.genome.fa \

# Genome annotation file
--gtf /scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/gencode.v32.annotation.gtf \

# Salmon will be doing all of the alignments and quantifications
--aligner star_salmon \

# letting pipeline know which type of annotation we are using
--gencode \

-- # PLESAE CHANGE TO YOUR EMAIL john.rinn@colorado.edu \
-c nextflow.config

date
```

Here is the run.sh in full form for nano adjustments
```{bash run.sh}

#!/bin/bash
#SBATCH -p long
#SBATCH --job-name=rna-protein-rnaseq
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=john.rinnn@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=6gb
#SBATCH --time=100:00:00
#SBATCH --output=nextflow.out
#SBATCH --error=nextflow.err

pwd; hostname; date
echo "You've requested $SLURM_CPUS_ON_NODE core."

module load singularity/3.1.1

nextflow run nf-core/rnaseq -r 3.4 \
-resume \
-profile singularity \
--input /scratch/Shares/rinn/JR/rna_protein_complexes/rnaseq/design.csv \
--fasta /scratch/Shares/rinn/genomes/Mus_musculus/Gencode/M25/GRCm38.p6.genome.fa \
--gtf /scratch/Shares/rinn/genomes/Mus_musculus/Gencode/M25/gencode.vM25.annotation.gtf \
--aligner star_salmon \
--gencode \
--email john.rinn@colorado.edu \
-c nextflow.config

date

```


Now we need the design file, similar to nf-core/chipseq it has to be very specific :)

sample,replicate,fastq_1,fastq_2,strandedness
Our data is not stranded, but this is an important consideration even if you don't have it.

This is an example of good reproducibility. We made the design file in 17 from code someone else could
run and get the exact same design file !


#TODO set up place to run and download fastqs from 17


