---
title: "18_Rnaseq_part_II"
author: "JR"
date: "11/12/2021"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(tximport)
library(DESeq2)
library(ggrepel)
library(pheatmap)
```

Let's do some RNAseq !

Goal: Align reads and get the number of reads in each gene in the genome using nf-core/rnaseq 

To use this new version of the nf-core/rnaseq pipeline (v3.4) 
So what we need is

design.csv
nextflow.config file
run.sh


Let's start with the config file -- very similar to nf-core/chipseq


```{bash}
process {
  executor='slurm'
  queue='short'
  memory='16 GB'
  maxForks=10
}
```

Now the run.sh file that gives nf-core/rnaseq all the instructions for nextflow to task. 
#TODO fix this !
```{bash}
#!/bin/bash
#SBATCH -p long
#SBATCH --job-name=HEPG2_rna_seq
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=john.rinn@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=6gb
#SBATCH --time=20:00:00
#SBATCH --output=nextflow.out
#SBATCH --error=nextflow.err


## This is the same as nfcore/chipseq start
pwd; hostname; date
echo "Here we go You've requested $SLURM_CPUS_ON_NODE core."

module load singularity/3.1.1

## Here we are readching out to the nf-core/rnaseq github to get the needed software
nextflow run nf-core/rnaseq -r 3.0 \
-resume \
-profile singularity \

## Here we are providing the paired ends to be merged and replicates etc...

--input design.csv

## There are different quantification methods (e.g., RSEM) we are choosing SALMON
--aligner star_salmon

## We need a genome
--fasta /scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/GRCh38.p13.genome.fa \
## We need a genome annotation
--gtf /scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/gencode.v32.annotation.gtf \

## SALMON needs to know some details on how the GTF is formatted. REFSEQ if a default so
## if we deviate from that we need to let SALMON know which annotation we are using.
--gencode \

## Standard stuff
--email john.rinn@colorado.edu \
-c nextflow.config

date


```

Here is the run.sh in final form
```{bash}
#!/bin/bash
#SBATCH -p long
#SBATCH --job-name=HEPG2_rna_seq
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=john.rinn@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=6gb
#SBATCH --time=20:00:00
#SBATCH --output=nextflow.out
#SBATCH --error=nextflow.err

pwd; hostname; date
echo "Here we go You've requested $SLURM_CPUS_ON_NODE core."

module load singularity/3.1.1

nextflow run nf-core/rnaseq -r 3.0 \
-resume \
-profile singularity \
--input design.csv
--aligner star_salmon
--fasta /Shares/rinn_class/data/genomes/human/gencode/v32/GRCh38.p13.genome.fa \
--gtf /Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf \
--gencode \
--email john.rinn@colorado.edu \
-c nextflow.config

date

```


Now we need the design file, similar to nf-core/chipseq it has to be very specific :)

sample,replicate,fastq_1,fastq_2,strandedness

Our data is not stranded, but this is an important consideration even if you don't have it.
