---
title: "19_Rnaseq_diff_expression"
author: "JR"
date: "11/20/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(tximport)
library(DESeq2)
library(ggrepel)
library(pheatmap)

source("../../../util/plotting_functions.R")
source("../../../util/_setup.R")
```

Today we will performing differential expression on HEPG2 RNAseq data. We will focus on comparing nuclear versus cytoplasmic RNAs. We can use this data to find genes that are nuclear or cytoplasmic enriched. We can use this data to determine if the high-binding promoters are biased for nuclear or cytoplasmic RNAs. Meaning the number of binding events indicates if the RNA product is sent to the cytoplasm (typically) or the nucleus. We can also use this data to determine if lncRNAs or mRNAs have similar or different nuclear/cytoplasmic localizations.


# Reading in samplesheet form 17_.Rmd
First let's read in th sample sheet we made in 17

```{r import sample sheet}
samplesheet <- read_csv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/17_API_RNASEQ/samplesheet.csv")

# ENCODE swapped labels on one sample !! 

# We of course need to swap the labels for total and insoluble cytoplasmic
# From looking in the browser at genes that should be nuclear: NEAT1 for ex.
# We have concluded that 
# hepg2_R2 -- is whole cell / total
# hepg2_insoluble_cytoplasmic_fraction_R2 -- is whole cell / total
# hepg2_R1 -- is insoluble_cytoplasmic
# hepg2_insoluble_cytoplasmic_fraction_R1 -- is insoluble_cytoplasmic

# Long story let's just fix it.

# changing the condition column to insoluble 
# note operation assignment comes last to just change text.

samplesheet[which(samplesheet$sample_name == "hepg2_R1"), "condition"] <- "insoluble_cytoplasmic_fraction"

# same for changing condition to total RNA
samplesheet[which(samplesheet$sample_name == "hepg2_insoluble_cytoplasmic_fraction_R2"), "condition"] <- "total"

# saving this to sample_sheets folder

write.csv(samplesheet, "corrected_samplesheet.csv")

```

# Loading in Gencode annotations
# also making gene_id to gene_symbol object 
Now let's read in the genome features also in our data folder
While we are at it, let's also print these out to make this go quicker in the future.
```{r import genome features}

# Importing genome annotations
gencode_gtf <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/gencode.v32.annotation.gtf")
# that took awhile let's make genome annotation files that can be quickly read in the future


# Let's reduce this big GRanges list to just genes
genes <- gencode_gtf[gencode_gtf$type == "gene"]


# let's make an object of gene_id and names
# we just saw above we can get both gene_id and name from 'genes'
# We can convert to DF for readability.

g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

write.csv(g2s, "g2s.csv")

# Nice now we have a handy conversion table of gene_id & gene_name
```

# reading in salmon merged gene counts.
Ok, next step is to read in GENE COUNTS (not TPM) This was exported from the NF_CORE RNAseq pipeline in the salmon folder. We want "salmon_merged_gene_counts.csv

This is in our results folder from 18_.Rmd
```{r reading in salmon counts convert to matrix}

# DEseq requires a matrix of raw counts so the following
# will turn this from a data.frame to matrix 
# NOTE: now common practice always think about adding "names" after import.

counts <- read_csv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/18_running_RNAseq_NF_CORE/NF_CORE_RUN/results/salmon/salmon_merged_gene_counts.csv")

# adding rownames
counts <- column_to_rownames(counts, "gene_id") %>%
as.matrix()
  
# Nice now we have a matrix we need to import to DeSeq
# However it has all the samples so we need to organize a bit.
```

# Connecting samplesheet to counts

Making an rlog counts matrix for all samples -- can read in to start in future
```{R sub-selecting counts_matrix to data only in samplesheet }

# First let's put the counts columns in the same order as the samplesheet
# If they are not then we would essentially be randomizing our data later!!

# REMOVING POLYA-selection SAMPLES
# NOTE: that right now, this pipeline run contains polyA and total RNA-seq
# we can filter to just the total RNA-seq by taking only the columns
# that are in the samplesheet (n = 10)
counts <- counts[,samplesheet$sample_id]

# This is a handy check point to make sure the labels in 
# sample sheet and counts are similar

all(colnames(counts) == samplesheet$sample_id)

# Now we know they're in the same order, and we can change the names
# It is nicer to have more descriptive names.
colnames(counts) <- samplesheet$sample_name

```

# creating and exporting normalized counts

DeSeq needs things like:
1) numbers to be rounded to an integer 
2) Add "factor" levels to conditions being compared
3) filter out any genes that don't have any counts

```{R Getting counts organized for DeSeq }

# Deseq wants round numbers so we are going to round them here:

counts <- round(counts)

# now we want to get rid of all genes with 0 across all samples.

counts_filtered <- counts[rowSums(counts) > 1,]


# In order to cache this and read it in later we'll
# save it as an RDS
write_rds(counts_filtered, "counts_matrix/counts_matrix.rds")

# We want to factor the condition variable. 
# In this case it doesn't matter which comes first the "control"
samplesheet$condition <- as.factor(samplesheet$condition)

# Now that the samplesheet is ready to used in DESeq2, 
# we'll write it out for future use.
write_rds(samplesheet, "final_samplesheet.rds")

```

# Running DeSeq to create rlog normalzied counts!

Although a lot of normalization has already occured from
Salmon (where we got the counts) we now need to determine the 
distribution of expression values and normalize again to prevent
low abundance genes from contributing to highly significant differential 
expression. 

Let's officially run DeSeq2!
```{R essential set up to running DeSeq}

# Next we want to normalize these counts 
# mostly to account for low abundance and noisy genes
# We need to do this in four steps:

# (1) first run DESeq2 by creating a dds object.

dds <- DESeqDataSetFromMatrix(countData = counts_filtered,
                              colData = samplesheet,
                              design = ~ condition)

# (2) run DESeq2 on dds object

dds <- DESeq(dds)

# (3) arrive at rlog normalized counts table!
# bild = T means it will normalize with out the "design ~ condition"
# seems redundant but these are the steps :)

rlog_counts <- rlog(dds, blind = TRUE)

# (4) now we retrieve the values using assay (an accessor function into rlog_counts)

rlog_counts_matrix <- assay(rlog_counts)

# Now we can write this out and START from here in the future.
# create new folder 'rlog_count_matrices'

write_rds(rlog_counts_matrix, "RNA_seq_results/rlog_count_normalized_data/rlog_counts_all.rds")

# reading back in to be sure we can start here in the future.
rlog_counts_matrix <- read_rds("rlog_count_matrices/rlog_counts_all.rds")
```

# Differential Expression (nuclear versus cytoplasmic)
Above we got to normalized counts that are ready to compared across
conditions. Note we "factored" the sample sheet so each fraction has
a 'factor' value. So we can now factor nuclear and compare to cytoplasmic.

Let's specialize sample sheet we can use to compare nuc/cyto samples.

```{R nuclear and cytoplasmic samplesheet}

# TODO BUG HERE ON OUT !!
# Now we will filter the sample sheet to just nuclear and cytoplasmic

samplesheet_nuc_cyto <- samplesheet %>%
  filter(condition %in% c("homosapien_cytosolic_fraction", "homosapien_nuclear_fraction"))
# Nice now we have a reduced sample sheet that is well organized.

# Next, let's triple check that the counts data
# contains the counts for the samples in the reduced sample sheet 
samplesheet_nuc_cyto$sample_name %in% colnames(counts)
# Cool, the 4 samples in filtered are same as counts

# Again we need to think about "rownames"
# Note the rownames are numbers -- this is 'meta' information we need.
# To make rownames we will turn into data.frame and add rownames
samplesheet_nuc_cyto <- samplesheet_nuc_cyto %>%
  as.data.frame()

# IMPORTANT: adding rownames from nuc_cyto_samplesheet by indexing
rownames(samplesheet_nuc_cyto) <- samplesheet_nuc_cyto$sample_name
# Now if we look at this object we see rownames as sample_name - nice.

# NOTE: DeSeq wants the control as a factor (fct), currently (chr)
# The first factor level is the "control" and all the other levels
# will be compared back to it.

# Let's make the condition a factor:
samplesheet_nuc_cyto$condition <- factor(samplesheet_nuc_cyto$condition, 
                                         levels = c("nuclear_fraction",
                                                    "cytosolic_fraction"))


# let's write out this samplesheet for future use
write_rds(samplesheet_nuc_cyto, "samplesheet_nuc_cyto.rds")

```


# START POINT
All the work we have done above can be skipped by 
reading in the following files.
```{r START POINT}

# Read in the raw counts
counts <- read_rds("counts_matrix/counts_matrix.rds")

# All normalized counts 
rlog_counts_matrix <- read_rds("RNA_seq_results/rlog_count_normalized_data/rlog_counts_all.rds")

# full sample_sheet
samplesheet <- read_rds("final_samplesheet.rds")

# nuc/cyto sample sheet
samplesheet_nuc_cyto <- read_rds("samplesheet_nuc_cyto.rds")


# finally a gene_id and gene_names object
g2s <- read.csv("g2s.csv")

```

# Heatmap of all samples
First let's cluster all the data to see how the samples relate.
```{r heatmap of all data}

# Now we can make a heatmap of all the different fractions

# First we will filter to genes that are variable across data.
hist(rowVars(rlog_counts_matrix))

# Filterting out genes that don't have variance of more than 2
rlog_var_genes_all <- rlog_counts_matrix[rowVars(rlog_counts_matrix) > 2,]

# row center the counts -- we need to flip the matrix
# we can only scale cols so we need to transmute (t)
# then turn it back with (t)
scaled_counts <- t(scale(t(rlog_var_genes_all))) %>%
  as.matrix()

# make heat map
pheatmap(scaled_counts, show_rownames = FALSE)

# saving
pdf("figures/all_vs_total_genes_heatmap.pdf")
pheatmap(scaled_counts, show_rownames = FALSE)
dev.off()
```

# Principal component analysis
Now let's use Principle Component Analysis to determine sample similarity
```{r PCA analysis}

# The first thing that we'll want to do is make a PCA plot.
# 'prcomp()' is the principle component analysis function.

?prcomp

# let's calculate the values associated with 'prcomp'
# install.packages("ggrepel") to move text so it doesn't overlap

rld_pca <- prcomp(t(rlog_counts_matrix))
rld_prcomps <- rld_pca$x %>% as.data.frame() %>%
  # note that x is in rld_pca as the PCA values
  rownames_to_column("sample_name") %>%
  select(sample_name, PC1, PC2)
rld_prcomps <- merge(samplesheet, rld_prcomps)


ggplot(rld_prcomps, aes(x = PC1, y = PC2, color = condition, label = replicate)) +
geom_point() +
geom_text_repel()

# RESULT: PCA plot looks like replicates are more similar than fraction type!

# Let's print this out

ggsave("figures/PCA_plot.pdf")

```

# Nuclear versus cytoplasmic differential expression
Next we will reduce out counts_matrix to just the values of nuc/cyto

```{r nuc/cyto  counts}

# first reduce the counts matrix to just nuc and cyto samples.
counts_nuc_cyto <- counts[, samplesheet_nuc_cyto$sample_name]

# This is a good way to triple check everything is in order.

stopifnot(all(rownames(samplesheet_nuc_cyto) == colnames(counts_nuc_cyto)))
# Great this evaluated to true and thus moved on -- we would get an error message otherwise.

```

# SwrrinG samplesheet for nuc vs cyto DEseq2
We are almost ready, last things to check:

1) DeSeq2 requires integer data 

2) Deseq2 needs the condition being compared to as a "factor" 
-- this is critical --

```{R organizing nuc/cyto counts and samplesheet for deseq}

# Ok now we need to turn the counts into integer data (int)

counts_nuc_cyto <- round(counts_nuc_cyto)
# This will round the counts, but still need to change to integer

mode(counts_nuc_cyto) <- "integer"
# We just coerced the data to integer and can now see that in env (int).


# Let's fileter to genes that have at least one count across cols.
nuc_cyto_counts_filtered <- counts_nuc_cyto[rowSums(counts_nuc_cyto) > 1,]

```

# Differential expression of nuc -vs- cyto DEseq2
Now let's look deeper into the results held in dds object
```{r DEseq2 nuc versus cyto}

# IMPORTANT: always double check organization of input files to DESeq!
# If this is not true the results will be a mix of wrong samples!
colnames(nuc_cyto_counts_filtered) == rownames(samplesheet_nuc_cyto)

# looks good, now let's make the required DEseq raw counts
dds <- DESeqDataSetFromMatrix(countData = nuc_cyto_counts_filtered,
                              colData = samplesheet_nuc_cyto,
                              design = ~ condition)


dds <- DESeq(dds)

# now we need to extract the results from dds:
res <- results(dds, name = "condition_cytosolic_fraction_vs_nuclear_fraction")

# It's easier to view it as a data.frame so we'll convert it.
# While we are at it we merge in gene_symbols (g2s)

res_df <- res %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s)

# Nice now we have a DF with a lot of good info for each gene.
```

QC plot: Volcano
```{r valono, MA and other plots}

# A first check is a "volcano" plot:
# x-axis = log fold change
# y-axis = pval

# Plotting volcano plot of nuclear vs cytoplasmic

ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), color = -log10(res_df$padj) > 90)) + 
  geom_point()

# saving volcano plot

ggsave("figures/nuc_cyto_volcano.pdf")

```

```{R MA PLOT}
# TODO base-mean seems to be an issue!!!!

# The 'MA plot' shows how low abundance affects can be an issue.
# x-axis = log(expression level)
# y-axis = log fold change

ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + 
  geom_point()

# This is a good QC plot and shows the fold changes
# are not biased to low "base-mean" levels

# saving MA PLOT

ggsave("figures/MA_plot_nuc_cyto.pdf")

```

let's see the range of pvalues in nuc-vs-cyto:
```{R pvalues nuc-vs-cyto}

hist(res_df$padj)

# how many genes above 0.05 padj?

num_sig <- res_df$padj < 0.05
summary(num_sig)

# We see about 14,268 genes exhibit differential expressed.
```


# DESeq with all subcellular fractions
```{r organizing sample sheet for comparison to total RNA for each fraction}

# We will use total RNA (whole cell) condition to compare everything back to.

# we will want to set the factor levels with "total" first
samplesheet$condition <- factor(samplesheet$condition, levels = c("total", "membrane_fraction", "insoluble_cytoplasmic_fraction", "cytosolic_fraction", "nuclear_fraction"))


# creating dds based on the factors above

dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = samplesheet,
                              design = ~ condition)


# Run the DESeq stats
dds <- DESeq(dds)


```


Let's look a little deeper into the dds results from total vs all
```{r looking at results from dds }

# We now have a bunch more results embedded in the dds object
resultsNames(dds)

# Let's just look at one of the results
res <- results(dds, name = "condition_membrane_fraction_vs_total")

# We can further index the information for just this sample:

res_df <- res %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_membrane_fraction_vs_total")

# Set the x-axes the same.
summary(res_df$log2FoldChange)

```


Ok that was a lot of RNAseq !!
Let's save the results:
```{r compiling all the results}

# A good starting point would be to use a for loop to make a data.frame 
# with all the results

results_names <- resultsNames(dds)

# let's look
results_names

# We don't care about the intercept, so we can leave that out
results_names <- results_names[-1]

# First we set up a data frame for logFC results.

res_df <- data.frame("gene_id" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name" = character(),
                     "result_name" = character())

# in parallel we can make the same results with "shrunken"
# logFC this normalizes low expressed genes to be less significant.

res_shrunken_df <- data.frame("gene_id" = character(), 
                              "baseMean" = numeric(), 
                              "log2FoldChange" = numeric(), 
                              "lfcSE" = numeric(),
                              "stat" = numeric(),
                              "pvalue" = numeric(),
                              "padj" = numeric(),
                              "gene_name" = character(),
                              "result_name" = character())


for(i in 1:length(results_names)) {
  results_name <- results_names[i]
  res <- results(dds, name = results_name)
  res_shrunken <- lfcShrink(dds, coef = results_name,  res = res)
  
  tmp_res_df <- res %>% as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  
  
  tmp_res_shrunken_df <- res_shrunken %>% as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  
  # Append to full data.frame
  res_df <- bind_rows(res_df, tmp_res_df)
  res_shrunken_df <- bind_rows(res_shrunken_df, tmp_res_shrunken_df)
}

# Let's save these res_df

write_rds(res_df, "results/deseq_results_df.rds")

write_rds(res_shrunken_df, "results/deseq_results_shrunken_lfc_df.rds")

```

Cool now we have two great data frames to move forward with.
Normally we would just move forward with shrunken logFC.
Let's add a couple more fun colums to this res_shrunken_df

```{R adding more information to res_shrunken_df }

# we could start here by reading in the files from above:
res_shrunken_df <- read_rds("results/deseq_results_shrunken_lfc_df.rds")

# First let's label the genes as being significant in any condition or not.
res_shrunken_df <- res_shrunken_df %>%
  group_by(gene_id) %>%
  mutate(sig = ifelse(any(padj < 0.05), "sig", "ns"))

# Let's clean up the column names a little bit
res_shrunken_df <- res_shrunken_df %>%
  mutate(subcellular_fraction = gsub("condition_|_fraction_vs_total", "", result_name))


sig_res_shrunked_df <- res_shrunken_df %>%
  filter(sig == "sig")

lfc_matrix <- sig_res_shrunked_df %>% 
  dplyr::select(gene_id, log2FoldChange, subcellular_fraction) %>% 
  pivot_wider(names_from = "subcellular_fraction", values_from = "log2FoldChange") %>%
  column_to_rownames("gene_id") %>%
  as.matrix()

pheatmap::pheatmap(lfc_matrix, show_rownames = FALSE, breaks = seq(-3, 3, length.out = 100))


```

