---
title: "20_binding_vs_expression"
author: "JR"
date: "11/21/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(GenomicRanges)
library(ggpubr)
source("../util/intersect_functions.R")
source("../util/_setup.R")
```


First let's read in the sample sheet to know what is what
```{r read in sample_sheet}

samplesheet <- read_rds("/Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/19_differential_expression/sample_sheets/final_samplesheet.rds")

```


Now let's read in the TPMs from salmon
```{r reading in salmon Salmon TPMs}

# reading in salmon counts TPM (note we previously read in counts)

salmon_tpm <- read.csv("/scratch/Shares/rinnclass/CLASS_2022/data/rnaseq/results/salmon/salmon_merged_gene_tpm.csv")

# Taking the mean TPM of replicate values for each subcellular fraction
# while we are at it we will add more intuitive row and col names 

#TODO understand this fully -- good example
tpm <- salmon_tpm %>% 
  pivot_longer(cols = 2:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samplesheet) %>%
  group_by(gene_id, condition) %>%
  summarize(tpm = mean(tpm, na.rm = T)) %>%
  pivot_wider(names_from = condition, values_from = tpm, names_prefix = "tpm_")
```

Now let's read in the promoter features data:
how many DBPs are bound at each promoter
```{r}
promoter_features_df <- read.csv("/Shares/rinn_class/data/CLASS_2022/class_exeRcises/analysis/13_ChIP_peak_features/peak_occurence_dataframe.csv")

# Now we can merge in the TPM data to this data.frame
promoter_features_df <- merge(promoter_features_df, tpm)

# saving this file
write.csv(promoter_features_df, "results/promoter_feature_df_20_tpm.csv")

# let's plot the number of DBPs bound as a densuty plot.

ggplot(promoter_features_df, aes(x = number_of_dbp)) +
  geom_density() +
  ggsave("figures/DBP_binding_density_plot.pdf")

# Remember this plot :) !

```
RESULT: There is a bimodal distriubution of the number of DBPs bound
to a given promoter. One that falls off around 100 DBPs and another
that emerged at ~250 DBPs on a promter. That's a lot.



Plotting the abundance of genes in each cellular fraction.
```{r TPM of genes in each fraction}

# First we need to the tpm DF into a matrix
tpm_matrix <- tpm %>% 
  column_to_rownames("gene_id") %>%
  as.matrix()

# And z-scale each row.
tpm_scaled <- t(scale(t(tpm_matrix)))

# We want to be sure that we're not removing rows that have NA in only one fraction.
# And right now we're not sure.
summary(is.na(tpm_matrix))

# NA = false, cool, but if there were we would want to remove them with complete.cases:
?complete.cases

# clearing all possible NAs
tpm_scaled <- tpm_scaled[complete.cases(tpm_scaled),]

pheatmap::pheatmap(tpm_scaled, show_rownames = FALSE)

# saving heatmap:
pdf("figures/tpm_all_fraction_genes_heatmap.pdf")
pheatmap::pheatmap(tpm_scaled, show_rownames = FALSE)
dev.off()

```
RESULT: 
(1) Most RNAs are abundant in the nucleus
(2) Some RNAs expressed in total that are not in other fractioins?


Now let's examine how binding effects expression.
We have published previously that the more binding
events at a promoter the more abundant the expression is.
Let's see if this holds for this larger number of DBPs
```{r}

# We will plot:
# x-axis = number of DBPs bound on the promoter
# y-axis = expression level of that gene.

ggplot(promoter_features_df, 
            aes(y = log2(tpm_total + 0.001), x = number_of_dbp, color = gene_type)) + 
  geom_point(data = promoter_features_df %>% filter(tpm_total < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) +
  # saving figure
  ggsave("figures/expression-vs-promoter_binding.pdf")


# The ones with high number of DBPs and no expression are our reservoirs
```
RESULTS:
(1) Expression is typically linearly correlatd with the number of DBPs at a proomter.
This validates known data from Mele et al. on a much larger scale.

(2) Linear correlation of DBP to expression level is similar for lncRNA and mRNA.
This is also validates that lncRNAs are typically lower abundant. 

(2) There is a very interestingn line on the bottom of the plot !
This means that there is a unique distribution of promoters that
are bound by many DBPs (up to 400!).

Let's see if the binding versus expression holds in the nuclear fraction
```{r binding versus nuclear expression}

ggplot(promoter_features_df, 
            aes(y = log2(tpm_nuclear_fraction + 0.001), x = number_of_dbp, color = gene_type)) + 
  geom_point(data = promoter_features_df %>% filter(tpm_total < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Nuclear Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) +
  # saving figure
  ggsave("figures/nuclear_expression-vs-promoter_binding.pdf")


```


Next we will determine the DBP binding versus cytoplasmic expression
```{Binding versus cytoplasmic expression}

ggplot(promoter_features_df, 
            aes(y = log2(tpm_cytosolic_fraction + 0.001), x = number_of_dbp, color = gene_type)) + 
            # note if we take off the log2 everything is very lowly expressed
  geom_point(data = promoter_features_df %>% filter(tpm_total < 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Cytoplasmic Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) +
  # saving figure
  ggsave("figures/cytoplasmic_expression-vs-promoter_binding.pdf")

```
RESULTS:
(1) everything seems to be low abundance 
(2) Some mRNAs are expressed in the nucleus -- we could look at this more later.
(3) The same linear trend holds but is driven by mostly low expression events.

Next we will directly test the lncRNA vs mRNA expression levels in total RNA. 
```{r determining lncRNA and mRNA expression levels in total RNA}

# We can see this from the trend lines that lncRNAs are less expressed,
# but here's another way of looking at that.
ggplot(promoter_features_df, aes(x = log2(tpm_total + 0.01), color = gene_type))+
  geom_density()

```
RESULT:
(1) This yet again confirms lncRNAs have lower expression levels than mRNAs.



We have previously observed that k562 cells also exhibit high binding promoters
that are not expressed. We termed them 'reservoirs' as they are a reservoir
for many Dna-protein interaction sites. Based on the plot above we observed that
this phenomena also exists in hepG2 cells as well. 

Based on this we next wanted to identify the 'reservoirs' in hepG2 cells.
```{r defining HEPG2 reservoirs}

# first we will use a cutoff of 100 DBPs.
# We can think about this more in the future.

promoter_features_df$hepg2_reservoir <- 
  as.numeric(promoter_features_df$number_of_dbp > 100 & 
               promoter_features_df$tpm_total < 0.001)

table(promoter_features_df$hepg2_reservoir)

```
RESULT:
(1) There are 628 reservoirs in HEPG2.



Now that we have defined reservoirs in hepG2 cells, we next want to determine how many 
are similar genomic regions in k562 and hepG2.
```{r reading in K%62 reservoirs}

k562_df <- read_csv("2020_k562_promoter_peak_df.csv")
# saving for future use this was hard to find :)
write_csv(k562_df, "2020_k562_promoter_peak_df.csv")

# next we want to merge the k562 adn Hepg2 DFs 
# first we should label what is k562 and what is Hepg2


# renaming cols to have k562



# To compare these reservoirs to this year's dataset, we can just merge in the relevant columns
k562_df <- k562_df %>% 
  dplyr::select(gene_id, reservoir, conservative_reservoir, tpm, expression, tf_binding, promoter_mean_tpm, promoter_median_tpm, promoter_max_tpm) %>%
  dplyr::rename(k562_reservoir = reservoir, 
                k562_conservative_reservoir = conservative_reservoir,
                k562_expression = expression,
                k562_tpm = tpm,
                k562_tf_binding = tf_binding,
                k562_promoter_mean_tpm =  promoter_mean_tpm,
                k562_promoter_median_tpm = promoter_median_tpm,
                k562_promoter_median_tpm = promoter_median_tpm,
                k562_promoter_max_tpm = promoter_max_tpm)

write.csv(k562_df,"k562_df.csv")

# TODO rename Hepg2

promoter_features_df <- merge(promoter_features_df, k562_df)


# Make a table of reservoir status
res_status <- promoter_features_df %>% 
  group_by(hepg2_reservoir, k562_reservoir, k562_conservative_reservoir) %>%
  summarize(count = n())
```

```{r}
# We can now write these out for safekeeping / use in other analyses
write_csv(promoter_features_df, "results/promoter_features_df.csv")
write_csv(tpm, "results/mean_tpm_per_condition.csv")
write_csv(samplesheet, "results/samplesheet.csv")
```

There's a lot to explore here. We have some potential analyses that you can pursue, or your group can decide on a direction that you're interested in.

1. How does the binding vs. expression change for the high binding promoters (> ~300 DBPs) vs low binders? And does this trend change across subcellular fractions?
2. For the genes that are exclusively in one fraction (i.e. exclusively nuclear) what do their promoter binding profiles have in common?
3. Are there any mRNAs that are exclusively in the nuclear fraction?
4. What does the expression of the K562 reservoirs look like in HEPG2 in terms of expression? Are they more lowly expressed? Or does their trend follow the binding vs expression trend for non-K562 reservoirs?
